---
title: "upbm: Analysis of Protein Binding Microarrays in R"
author: Patrick K. Kimes
date: "`r BiocStyle::doc_date()`"
package: "`r BiocStyle::pkg_ver('upbm')`"
abstract: >
  "Protein binding microarrays (PBMs) allow for the high-throughput quantification of DNA binding specificities of proteins, e.g. transcription factors, in vitro. While several probe designs have been proposed for PBMs, universal-design PBMs (uPBMs) which cover all 10-mer DNA sequences, have enjoyed the most wide-spread use and success. The *upbm* package provides tools and methods for organizing probe-level PBM data from raw GPR scan files, summarizing PBM data at k-mer resolution, and performing statistical inference when replicate experiments are available. The package was developed for the analysis of uPBMs, but should be appropriate for most other PBM designs as well. upbm package version: `r packageVersion("upbm")`"
output:
  BiocStyle::html_document:
    highlight: pygments
    toc: true
    fig_width: 12
    fig_height: 4
vignette: >
  %\VignetteIndexEntry{upbm: introduction}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

```{r echo=FALSE, include=FALSE}
knitr::opts_chunk$set(tidy = FALSE, cache = TRUE, dev = "png",
                      message = FALSE, error = FALSE, warning = TRUE)
```

# Introduction

While a large collection of methods and software packages have been developed for the processing and analysis of traditional DNA microarrays, the development of tools for protein binding microarrays (PBMs) has been limited. Due to the substantial difference of what is being measured by PBMs (protein-DNA binding) and DNA microarrays (nucleic acid hybridization), blindly applying DNA microarray tools to PBM data can lead to unexpected problems. 

This package provides tools and methods for reading, plotting, and analyzing protein binding microarray (PBM) data in *R* using standard [*Bioconductor*](https://www.bioconductor.org/) [data structures](http://bioconductor.org/developers/how-to/commonMethodsAndClasses/). The primary focus is on universal-design PBMs (uPBMs) which provide approximately uniform coverage of all 8-mer DNA sequences across probes on the array [@berger2006].

# Installation

To install the latest version of *upbm* along with two dependencies, *upbmAux* and *upbmData*, use the following command from the [`BiocManager`](https://CRAN.R-project.org/package=BiocManager) package:

```{r, eval = FALSE}
BiocManager::install("pkimes/upbm")
BiocManager::install("pkimes/upbmAux")
BiocManager::install("pkimes/upbmData")
```

After installing, the package can be loaded to make the functions available in *R*.

```{r}
##suppressPackageStartupMessages(library("upbm"))
devtools::load_all()
```

# HOXC9 Dataset

Data from a triplicate series of PBM experiments are used to illustrate the features of the *upbm* package in this vignette. The dataset includes GenePix Results (GPR) data from Alexa488 scans for wild type human transcription factor (TF) HOXC9 and three allelic variants (K195R, R193K, R222W). Corresponding Cy3 scans are also included for all arrays. Each array was scanned for Alexa488 fluorescence at multiple PMT gains and Cy3 at a single PMT gain. The series comes from a larger set of experiments studying the effect of allelic variants on human transcription factor binding (manuscript in preparation). The HOXC9 data objects are stored in a separate *upbmData* package for convenience and include a table of sample metadata (`hoxc9table`), a *SummarizedExperiment* of Alexa488 scan data (`hoxc9alexa`), and a *SummarizedExperiment* of Cy3 scan data (`hoxc9cy3`).

```{r}
data(hoxc9table, package = "upbmData")
data(hoxc9alexa, package = "upbmData")
data(hoxc9cy3, package = "upbmData")
```

Metadata for Alexa488 and Cy3 scans are provided in the sample table.

```{r}
hoxc9table
```

Several important metadata variables are recorded in the table - notably, `scan` (scan type), `gpr` (path to the GPR file), `lp` (PMT gain of the scan), and `condition` (HOXC9 allele on the scanned array).

## Details

Details on how the raw GPR files for the HOXC9 data can be accessed and how the data was loaded to create the `hoxc9table`, `hoxc9alexa`, `hoxc9cy3` objects are described in the *upbmData* vignette.

# Quickstart

Using the example data included in the *upbmData* package, we first demonstrate how GPR data can be analyzed to obtain 8-mer-level affinity summaries without describing each step in too much detail. Additional information on how each 

Input data must be available as GenePix Results (GPR) files, with each GPR file corresponding to a single scan of a sample. The GPR files should be organized as a table of scans with each row detailing any important metadata about the scan (e.g. date and scan parameters) and sample(e.g. the name of the transcription factor). The path to the GPR file **must** be specified in a `gpr` column of the table.

The table of scan information for this example dataset are provided as `hoxc9table`. The scan files listed in the `gpr` column can be read in using the `buildPBMExperiment(..)` function. If possible, a corresponding probe design file should be also be specified. Standard uPBM designs, such as the 8x60k design, are included in the *upbmAux* package.

```{r}
data(pbm_8x60k_v1, package = "upbmAux")
head(pbm_8x60k_v1)
```

The scans are now read in along with the specified probe design. An error will be thrown if the dimensions of the GPR files and the specified table of probes.

```{r, eval = FALSE}
hoxc9data <- buildPBMExperiment(hoxc9table, probes = pbm_8x60k_v1)
```

The data is read in as a single *SummarizedExperiment* object, with scans as columns and probes as rows. 




The data is now available for plotting and analysis. Here, we show an example of computing the spatial bias in a subset of samples, and plotting the bias across the grid of probe coordinates.

```{r}
pbm_adjusted <- spatiallyAdjust(pbmdat[, 1:3])
```

```{r, fig.width = 12, fig.height = 4}
pbmPlotChip(pbm_adjusted, assay_name = "spatialbias", log_scale = FALSE)
```

More details on the available plotting and analysis functions are provided in the sections below.







# Overview

## Reading GPR Data

## Plotting GPR Data

## Pre-Processing GPR Data

## Filtering GPR Data

## Quantifying GPR Data

## Testing GPR Data



















# Related Work

A separate set of tools developed in Perl for the analysis of universal-design PBMs (uPBMs) is available for download at the [*Universal PBM Analysis Suite* site](the_brain.bwh.harvard.edu/PBMAnalysisSuite/indexSep2017.html) [@berger2009]. The *upbm* package builds on the work of the *Universal PBM Analysis Suite*, and provides alternative summaries of uPBM data not developed as part of the original suite of tools.

A primary purpose of both software packages is the summarization of DNA binding affinities at the level of individual K-mers. To this end, the *Universal PBM Analysis Suite* summarizes K-mer binding affinity for individual PBM samples with two metrics: a non-parametric rank-based enrichment score (E-score), and the cross-probe median intensity (calculated across all probes containing the corresponding K-mer). In contrast, with *upbm*, K-mer binding affinities are estimated using a mixed effects model, with affinities effectively estimated as the weighted average log-transformed probe intensity across all probes containing the corresponding K-mer. This differs from the cross-probe median intensity in serveral important ways. First, the relative contribution of individual probes to the K-mer summary is inversely related to the variance of the probe across samples, downweighting highly variable probes. Second, to estimate probe-level variability, replicate samples are necessary. Third, because the variance structure is explicitly modeled, statistical testing and inference is possible at the level of K-mers.

In addition to the summarization of DNA binding affinities at the level of individual K-mers, the *upbm* package also allows for testing for differential affinity between groups, e.g. to contrast binding affinity across allelic variants of the same transcription factor. Additionally, the package also allows for identifying preferentially bound K-mers for each groups of replicate samples. The *Universal PBM Analysis Suite* provides the ability to construct mononucleotide position weight matrices (PWMs), again, for individual PBM samples. Approaches for testing differential affinity between samples using the output of the *Universal PBM Analysis Suite* have also been previously described [@barrera2016].

Many of the PBM pre-processing steps of the *Universal PBM Analysis Suite* are maintained in the *upbm* package. These steps have been re-implemented, with some modifications, e.g. in the process of Cy3 normalization, to enable and streamline the complete analysis of PBM data in *R*.

# Alternative Inputs

While PBM data are commonly stored as GPR scan files, several similar probe-level summaries files have been generated and made publicly available on platforms include [Uniprobe](http://the_brain.bwh.harvard.edu/uniprobe/) [@hume2014]. In addition to raw GPR data, the *upbm* software can be used to read in alternative sources of extrapolated or processed probe-level intensity data, e.g. from [Masliner](http://arep.med.harvard.edu/masliner/supplement.htm) or the *Universal PBM Analysis Suite*. We recommend using original GPR data files when possible.

# References

