<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>upbm: Inference Details • upbm</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="upbm: Inference Details">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">upbm</a>
        <span class="version label label-info" data-toggle="tooltip" data-placement="bottom" title="GitHub pre-release">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/upbm-quickstart.html">Quick Start</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">General</li>
    <li>
      <a href="../articles/upbm-quickstart.html">Quickstart Guide</a>
    </li>
    <li>
      <a href="../articles/upbm-classes.html">PBM Class Details</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Analysis</li>
    <li>
      <a href="../articles/analysis-preprocessing.html">Sample Pre-Processing</a>
    </li>
    <li>
      <a href="../articles/analysis-summarization.html">Probe and K-mer Summarization</a>
    </li>
    <li>
      <a href="../articles/analysis-inference.html">K-mer-Level Inference</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Other</li>
    <li>
      <a href="../articles/other-qcplots.html">Quality Control Plots</a>
    </li>
    <li>
      <a href="../articles/other-tidydata.html">Tidy PBM Data</a>
    </li>
  </ul>
</li>
<li>
  <a href="https://github.com/pkimes/upbm">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>upbm: Inference Details</h1>
                        <h4 class="author">Patrick K. Kimes</h4>
            
            <h4 class="date">15 February 2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/pkimes/upbm/blob/master/vignettes/analysis-inference.Rmd"><code>vignettes/analysis-inference.Rmd</code></a></small>
      <div class="hidden name"><code>analysis-inference.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>For an introduction to the <em>upbm</em> package, please see the quick start vignette (<code><a href="../articles/upbm-quickstart.html">vignette("upbm-quickstart")</a></code>). Here, we provide details on the various testing procedures briefly referenced in the quick start vignette. We assume basic knowledge of the <em>PBMExperiment</em> class (see <code><a href="../articles/upbm-classes.html">vignette("upbm-classes")</a></code>). The same example HOXC9 dataset from the <em>upbmData</em> package is also used in this vignette to illustrate the various functions.</p>
<p>The <em>upbm</em> package includes three testing functions which all take the output of <code><a href="../reference/kmerFit.html">upbm::kmerFit</a></code> and return a table of 8-mer-level effect sizes and significance estimates. In the quick start vignette, only a single testing procedure was covered, namely, differential affinity testing using <code><a href="../reference/kmerTestContrast.html">upbm::kmerTestContrast</a></code>. The three testing procedures are:</p>
<ul>
<li>preferential affinity testing (<code><a href="../reference/kmerTestAffinity.html">upbm::kmerTestAffinity</a></code>),</li>
<li>differential affinity testing (<code><a href="../reference/kmerTestContrast.html">upbm::kmerTestContrast</a></code>),</li>
<li>differential specificity testing (<code><a href="../reference/kmerTestSpecificity.html">upbm::kmerTestSpecificity</a></code>).</li>
</ul>
<p>We cover each procedure in detail in this vignette.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">"upbm"</span>))</span></code></pre></div>
<p>Throughout, we will be using <code><a href="https://rdrr.io/pkg/broom/man/reexports.html">broom::tidy</a></code> to tidy data stored in <em>PBMExperiment</em> and <em>SummarizedExperiment</em> objects for interactive analysis and plotting (see <code><a href="../articles/other-tidydata.html">vignette("other-tidydata")</a></code>). We load the <em>dplyr</em> package to help with interactive analysis.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="kw"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">"dplyr"</span>))</span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="kw"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">"ggplot2"</span>))</span></code></pre></div>
</div>
<div id="hoxc9-dataset" class="section level1">
<h1 class="hasAnchor">
<a href="#hoxc9-dataset" class="anchor"></a>HOXC9 Dataset</h1>
<p>For details on the example HOXC9 dataset, see the quick start vignette in this package or the <em>upbmData</em> package vignette. Here, we will use both Alexa488 and Cy3 scans.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(hoxc9alexa, <span class="dt">package =</span> <span class="st">"upbmData"</span>)</span>
<span id="cb3-2"><a href="#cb3-2"></a><span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(hoxc9cy3, <span class="dt">package =</span> <span class="st">"upbmData"</span>)</span></code></pre></div>
<p>Here, we again use just single PMT gain scan for the Alexa488 scans.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>alexa_subset &lt;-<span class="st"> </span>hoxc9alexa[, <span class="kw">colData</span>(hoxc9alexa)<span class="op">$</span>pmt <span class="op">==</span><span class="st"> </span><span class="dv">450</span>]</span></code></pre></div>
<p>We also load the reference Cy3 dataset.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(refcy3_8x60k_v1, <span class="dt">package =</span> <span class="st">"upbmAux"</span>)</span></code></pre></div>
</div>
<div id="sample-processing" class="section level1">
<h1 class="hasAnchor">
<a href="#sample-processing" class="anchor"></a>Sample Processing</h1>
<p>We will perform the earlier steps of pre-processing using the <code>upbm::pmbPreprocess</code> wrapper. Details on pre-processing steps are covered in an earlier vignette (see <code>vignette("analysis-preprocessing"</code>).</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a>alexa_subset &lt;-<span class="st"> </span><span class="kw"><a href="../reference/upbmPreprocess.html">upbmPreprocess</a></span>(<span class="dt">pe =</span> alexa_subset,</span>
<span id="cb6-2"><a href="#cb6-2"></a>                               <span class="dt">cy3pe =</span> hoxc9cy3,</span>
<span id="cb6-3"><a href="#cb6-3"></a>                               <span class="dt">cy3refpe =</span> refcy3_8x60k_v1,</span>
<span id="cb6-4"><a href="#cb6-4"></a>                               <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</span></code></pre></div>
<p>We will further summarize results to 8-mer-level affinity estimates with <code>HOXC9-REF</code> as the baseline condition. As above, details on summarization steps are covered in an earlier vignette (see <code>vignette("analysis-summarization"</code>).</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>alexa_pfit &lt;-<span class="st"> </span><span class="kw"><a href="../reference/probeFit.html">probeFit</a></span>(alexa_subset)</span>
<span id="cb7-2"><a href="#cb7-2"></a>alexa_kfit &lt;-<span class="st"> </span><span class="kw"><a href="../reference/kmerFit.html">kmerFit</a></span>(alexa_pfit, <span class="dt">baseline =</span> <span class="st">"HOXC9-REF"</span>)</span></code></pre></div>
</div>
<div id="preferential-affinity" class="section level1">
<h1 class="hasAnchor">
<a href="#preferential-affinity" class="anchor"></a>Preferential Affinity</h1>
<p>In contrast to the two other testing procedures which consider <strong>differential</strong> behavior across conditions, preferential affinity considers each condition separately. For each condition, we apply a testing procedure to determine the set of 8-mers <em>preferentially</em> bound by the condition. The <code><a href="../reference/kmerTestAffinity.html">upbm::kmerTestAffinity</a></code> function only takes the single <em>PBMExperiment</em> returned by <code><a href="../reference/kmerFit.html">upbm::kmerFit</a></code> as input.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a>alexa_pa &lt;-<span class="st"> </span><span class="kw"><a href="../reference/kmerTestAffinity.html">kmerTestAffinity</a></span>(alexa_kfit)</span>
<span id="cb8-2"><a href="#cb8-2"></a>alexa_pa</span></code></pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 32896 4 
## metadata(0):
## assays(6): affinityEstimate affinityVariance ... affinityP
##   affinityQ
## rownames: NULL
## rowData names(1): seq
## colnames(4): HOXC9-K195R HOXC9-R193K HOXC9-R222W HOXC9-REF
## colData names(0):</code></pre>
<p>The returned <em>PBMExperiment</em> object contains the results of affinity testing in several assays along with many of the same assays as the input.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a><span class="kw">assayNames</span>(alexa_pa)</span></code></pre></div>
<pre><code>## [1] "affinityEstimate" "affinityVariance" "affinitySignal"  
## [4] "affinityZ"        "affinityP"        "affinityQ"</code></pre>
<p>Importantly, the new <code>affinitySignal</code> assay contains the estimated “signal” component of the K-mer affinity (<code>affinityEstimate</code> assay). The signal component is computed by fitting a normal and expontential convolution to the K-mer affinity estimates and taking the expected exponential component. This is similar to a step included in the RMA approach to microarray data normalization <span class="citation">(Irizarry et al. 2003)</span>. The <code>affinityZ</code>, <code>affinityP</code> and <code>affinityQ</code>, contain the <em>z</em>-score, un-adjusted and adjusted <em>p</em>-values from testing the null hypothesis that <code>affinitySignal</code> is zero.</p>
<p>The distribution of <code>affinityEstimates</code> can be colored by the adjusted <em>p</em>-values (<code>affinityQ</code>) to show the distribution of preferentially bound 8-mers in each condition. We bin the significance values for cleaner plotting.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a>pa_dat &lt;-<span class="st"> </span>broom<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/broom/man/reexports.html">tidy</a></span>(alexa_pa, <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"affinityEstimate"</span>, <span class="st">"affinityQ"</span>))</span>
<span id="cb12-2"><a href="#cb12-2"></a></span>
<span id="cb12-3"><a href="#cb12-3"></a>pa_dat <span class="op">%&gt;%</span></span>
<span id="cb12-4"><a href="#cb12-4"></a><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">affinityQ =</span> <span class="kw"><a href="https://rdrr.io/r/base/cut.html">cut</a></span>(affinityQ, <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1e0</span>, <span class="fl">1e-2</span>, <span class="fl">1e-3</span>, <span class="fl">1e-6</span>, <span class="dv">0</span>),</span>
<span id="cb12-5"><a href="#cb12-5"></a>                                  <span class="dt">include.lowest =</span> <span class="ot">TRUE</span>),</span>
<span id="cb12-6"><a href="#cb12-6"></a>                  <span class="dt">affinityQ =</span> <span class="kw"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(affinityQ, <span class="dt">levels =</span> <span class="kw"><a href="https://rdrr.io/r/base/rev.html">rev</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/levels.html">levels</a></span>(affinityQ)))) <span class="op">%&gt;%</span></span>
<span id="cb12-7"><a href="#cb12-7"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> affinityEstimate, <span class="dt">fill =</span> affinityQ)) <span class="op">+</span></span>
<span id="cb12-8"><a href="#cb12-8"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span>(<span class="dt">binwidth =</span> <span class="fl">.1</span>, <span class="dt">boundary =</span> <span class="dv">10</span>, <span class="dt">color =</span> <span class="st">'black'</span>, <span class="dt">position =</span> <span class="st">"stack"</span>, <span class="dt">alpha =</span> <span class="dv">1</span>) <span class="op">+</span></span>
<span id="cb12-9"><a href="#cb12-9"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_fill_brewer</a></span>(<span class="st">"q-value"</span>, <span class="dt">palette =</span> <span class="st">"Spectral"</span>, <span class="dt">direction =</span> <span class="dv">-1</span>, <span class="dt">drop =</span> <span class="ot">FALSE</span>,</span>
<span id="cb12-10"><a href="#cb12-10"></a>                      <span class="dt">na.value =</span> <span class="st">"black"</span>) <span class="op">+</span></span>
<span id="cb12-11"><a href="#cb12-11"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() <span class="op">+</span></span>
<span id="cb12-12"><a href="#cb12-12"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>( <span class="op">~</span><span class="st"> </span>cname, <span class="dt">nrow =</span> <span class="dv">1</span>, <span class="dt">scales =</span> <span class="st">"free_y"</span>) <span class="op">+</span></span>
<span id="cb12-13"><a href="#cb12-13"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span>(<span class="st">"HOXC9 8-mer preferential affinity"</span>)</span></code></pre></div>
<p><img src="analysis-inference_files/figure-html/prefaff-plot-1.png" width="960"></p>
</div>
<div id="differential-affinity" class="section level1">
<h1 class="hasAnchor">
<a href="#differential-affinity" class="anchor"></a>Differential Affinity</h1>
<p>Differential affinity is likely to most intuitive of the three testing procedures. With differential affinity, we are interested in identifying the set of 8-mers with differential signal across PBM samples between the specified baseline and all other conditions. The underlying hypothesis we test here is whether the estimated affinity differences are zero or not.</p>
<p>The <code><a href="../reference/kmerTestContrast.html">upbm::kmerTestContrast</a></code>, again, only takes the single <em>PBMExperiment</em> returned by <code><a href="../reference/kmerFit.html">upbm::kmerFit</a></code> as input.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a>alexa_da &lt;-<span class="st"> </span><span class="kw"><a href="../reference/kmerTestContrast.html">kmerTestContrast</a></span>(alexa_kfit)</span>
<span id="cb13-2"><a href="#cb13-2"></a>alexa_da</span></code></pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 32896 4 
## metadata(0):
## assays(6): contrastAverage contrastDifference ... contrastP
##   contrastQ
## rownames: NULL
## rowData names(1): seq
## colnames(4): HOXC9-K195R HOXC9-R193K HOXC9-R222W HOXC9-REF
## colData names(0):</code></pre>
<p>The returned <em>PBMExperiment</em> object contains the results of differential affinity testing a few new assays. The <code>contrastZ</code>, <code>contrastP</code>, and <code>contrastQ</code> contain the <em>z</em>-score, and unadjusted and adjusted <em>p</em>-values for the test at each 8-mer.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a><span class="kw">assayNames</span>(alexa_da)</span></code></pre></div>
<pre><code>## [1] "contrastAverage"    "contrastDifference" "contrastVariance"  
## [4] "contrastZ"          "contrastP"          "contrastQ"</code></pre>
<p>Here, the <em>z</em>-score is simply the ratio of the <code>contrastDifference</code> and <code>contrastVariance</code> computed during the <code><a href="../reference/kmerFit.html">upbm::kmerFit</a></code> call. We can again plot the results as an MA plot and color 8-mers by significance level. We also remember to exclude the baseline condition, <code>HOXC9-REF</code>, from the table as all of the values are <code>NA</code>s.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a>da_dat &lt;-<span class="st"> </span>broom<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/broom/man/reexports.html">tidy</a></span>(alexa_da, <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"contrastAverage"</span>, <span class="st">"contrastDifference"</span>,</span>
<span id="cb17-2"><a href="#cb17-2"></a>                                  <span class="st">"contrastQ"</span>))</span>
<span id="cb17-3"><a href="#cb17-3"></a></span>
<span id="cb17-4"><a href="#cb17-4"></a>da_dat &lt;-<span class="st"> </span>da_dat <span class="op">%&gt;%</span></span>
<span id="cb17-5"><a href="#cb17-5"></a><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(cname <span class="op">!=</span><span class="st"> "HOXC9-REF"</span>) <span class="op">%&gt;%</span></span>
<span id="cb17-6"><a href="#cb17-6"></a><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">contrastQ =</span> <span class="kw"><a href="https://rdrr.io/r/base/cut.html">cut</a></span>(contrastQ, <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1e0</span>, <span class="fl">1e-2</span>, <span class="fl">1e-3</span>, <span class="fl">1e-6</span>, <span class="dv">0</span>),</span>
<span id="cb17-7"><a href="#cb17-7"></a>                                  <span class="dt">include.lowest =</span> <span class="ot">TRUE</span>),</span>
<span id="cb17-8"><a href="#cb17-8"></a>                  <span class="dt">contrastQ =</span> <span class="kw"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(contrastQ, <span class="dt">levels =</span> <span class="kw"><a href="https://rdrr.io/r/base/rev.html">rev</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/levels.html">levels</a></span>(contrastQ))))</span>
<span id="cb17-9"><a href="#cb17-9"></a></span>
<span id="cb17-10"><a href="#cb17-10"></a>da_dat <span class="op">%&gt;%</span></span>
<span id="cb17-11"><a href="#cb17-11"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> contrastAverage, <span class="dt">y =</span> contrastDifference,</span>
<span id="cb17-12"><a href="#cb17-12"></a>               <span class="dt">color =</span> contrastQ)) <span class="op">+</span></span>
<span id="cb17-13"><a href="#cb17-13"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>(<span class="dt">alpha =</span> <span class="dv">1</span><span class="op">/</span><span class="dv">10</span>) <span class="op">+</span></span>
<span id="cb17-14"><a href="#cb17-14"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_color_brewer</a></span>(<span class="st">"q-value"</span>, <span class="dt">palette =</span> <span class="st">"Spectral"</span>, <span class="dt">direction =</span> <span class="dv">-1</span>, <span class="dt">drop =</span> <span class="ot">FALSE</span>,</span>
<span id="cb17-15"><a href="#cb17-15"></a>                       <span class="dt">na.value =</span> <span class="st">"black"</span>) <span class="op">+</span></span>
<span id="cb17-16"><a href="#cb17-16"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span>(<span class="dt">color =</span> <span class="st">'gray'</span>, <span class="dt">yintercept =</span> <span class="dv">0</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb17-17"><a href="#cb17-17"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() <span class="op">+</span></span>
<span id="cb17-18"><a href="#cb17-18"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/guides.html">guides</a></span>(<span class="dt">color =</span> <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html">guide_legend</a></span>(<span class="dt">override.aes =</span> <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">alpha =</span> <span class="dv">1</span>))) <span class="op">+</span></span>
<span id="cb17-19"><a href="#cb17-19"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span>(. <span class="op">~</span><span class="st"> </span>cname) <span class="op">+</span></span>
<span id="cb17-20"><a href="#cb17-20"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span>(<span class="st">"HOXC9 8-mer differential affinity"</span>)</span></code></pre></div>
<p><img src="analysis-inference_files/figure-html/difaff-plot-1.png" width="960"></p>
<p>We can see that few 8-mers are called differential for most of the conditions. More formally, we can count up the number of 8-mers at each of the different significance thresholds above.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a>da_dat <span class="op">%&gt;%</span></span>
<span id="cb18-2"><a href="#cb18-2"></a><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/tally.html">count</a></span>(cname, contrastQ) <span class="op">%&gt;%</span></span>
<span id="cb18-3"><a href="#cb18-3"></a><span class="st">    </span>tidyr<span class="op">::</span><span class="kw"><a href="https://tidyr.tidyverse.org/reference/spread.html">spread</a></span>(contrastQ, n, <span class="dt">fill =</span> <span class="dv">0</span>)</span></code></pre></div>
<pre><code>## # A tibble: 3 x 5
##   cname       `(0.01,1]` `(0.001,0.01]` `(1e-06,0.001]` `[0,1e-06]`
##   &lt;chr&gt;            &lt;dbl&gt;          &lt;dbl&gt;           &lt;dbl&gt;       &lt;dbl&gt;
## 1 HOXC9-K195R      32885              7               3           1
## 2 HOXC9-R193K      32882             14               0           0
## 3 HOXC9-R222W      20623           2991            4075        5207</code></pre>
</div>
<div id="differential-specificity" class="section level1">
<h1 class="hasAnchor">
<a href="#differential-specificity" class="anchor"></a>Differential Specificity</h1>
<p>Differential specificity, while similar to affinity is treated separately. In some comparisons, the two proteins being contrasted may have completely different binding preferences, preferring different subsets of 8-mers. We call this phenomenon <em>differential specificity</em>.</p>
<p>In the MA plot above, we see that the general trend for all contrasts is a single trajectory, even for the R222W allele. This is the result of shared specificity, even with decreased affinity. In contrast, when differential specificity is present, rather than follow a single trajectory, the MA plot will appear to split, with 8-mers preferential in the two condition spreading.</p>
<p>Intuitively, testing for differential specificity is performed by first fitting a single trajectory to the MA plot, followed by testing for deviations from the trajectory. Since deviations from the baseline condition are often of more interest, rather than using the average between conditions as the x-axis, the trajectory is fit with the reference affinity along the x-axis. This can be changed by setting <code>useref = FALSE</code> in the call to <code><a href="../reference/kmerTestSpecificity.html">upbm::kmerTestSpecificity</a></code>.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a>alexa_ds &lt;-<span class="st"> </span><span class="kw"><a href="../reference/kmerTestSpecificity.html">kmerTestSpecificity</a></span>(alexa_kfit)</span>
<span id="cb20-2"><a href="#cb20-2"></a>alexa_ds</span></code></pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 32896 4 
## metadata(0):
## assays(9): contrastAverage contrastDifference ... specificityQ
##   specificityAxis
## rownames: NULL
## rowData names(1): seq
## colnames(4): HOXC9-K195R HOXC9-R193K HOXC9-R222W HOXC9-REF
## colData names(0):</code></pre>
<p>By default, the trajectory is fit using cubic B-splines with 3 knots placed evenly spaced along the range of the x-axis. Alternatively, a LOESS smoother can be fit by specifying <code>method = "loess"</code>.</p>
<p>The returned <em>PBMExperiment</em> includes several new assays.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a><span class="kw">assayNames</span>(alexa_ds)</span></code></pre></div>
<pre><code>## [1] "contrastAverage"    "contrastDifference" "contrastVariance"  
## [4] "contrastFit"        "contrastResidual"   "specificityZ"      
## [7] "specificityP"       "specificityQ"       "specificityAxis"</code></pre>
<p>The x-axis used for fitting the specificity trend is stored in the <code>specificityAxis</code> assay. The fit and residuals are included in the <code>contrastFit</code> and <code>contrastResidual</code> assays. Finally, the <code>specificityZ</code>, `<code>specificityP</code>, and <code>specificityQ</code> assays contain the <em>z</em>-score, unadjusted and adjusted <em>p</em>-values from testing the null hypothesis that <code>contrastResidual</code> is zero.</p>
<p>We can plot the data along with the <code>contrastFit</code> to see the B-spline trajectories fit to the data.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a>ds_dat &lt;-<span class="st"> </span>broom<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/broom/man/reexports.html">tidy</a></span>(alexa_ds, <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"specificityAxis"</span>, <span class="st">"specificityQ"</span>,</span>
<span id="cb24-2"><a href="#cb24-2"></a>                                  <span class="st">"contrastDifference"</span>, <span class="st">"contrastFit"</span>,</span>
<span id="cb24-3"><a href="#cb24-3"></a>                                  <span class="st">"contrastResidual"</span>))</span>
<span id="cb24-4"><a href="#cb24-4"></a></span>
<span id="cb24-5"><a href="#cb24-5"></a>ds_dat &lt;-<span class="st"> </span>ds_dat <span class="op">%&gt;%</span></span>
<span id="cb24-6"><a href="#cb24-6"></a><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(cname <span class="op">!=</span><span class="st"> "HOXC9-REF"</span>) <span class="op">%&gt;%</span></span>
<span id="cb24-7"><a href="#cb24-7"></a><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">specificityQ =</span> <span class="kw"><a href="https://rdrr.io/r/base/cut.html">cut</a></span>(specificityQ, <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1e0</span>, <span class="fl">1e-2</span>, <span class="fl">1e-3</span>, <span class="fl">1e-6</span>, <span class="dv">0</span>),</span>
<span id="cb24-8"><a href="#cb24-8"></a>                                  <span class="dt">include.lowest =</span> <span class="ot">TRUE</span>),</span>
<span id="cb24-9"><a href="#cb24-9"></a>                  <span class="dt">specificityQ =</span> <span class="kw"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(specificityQ, <span class="dt">levels =</span> <span class="kw"><a href="https://rdrr.io/r/base/rev.html">rev</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/levels.html">levels</a></span>(specificityQ))))</span>
<span id="cb24-10"><a href="#cb24-10"></a></span>
<span id="cb24-11"><a href="#cb24-11"></a>ds_dat <span class="op">%&gt;%</span></span>
<span id="cb24-12"><a href="#cb24-12"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> specificityAxis)) <span class="op">+</span></span>
<span id="cb24-13"><a href="#cb24-13"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>(<span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">y =</span> contrastDifference, <span class="dt">color =</span> specificityQ), <span class="dt">alpha =</span> <span class="dv">1</span><span class="op">/</span><span class="dv">10</span>) <span class="op">+</span></span>
<span id="cb24-14"><a href="#cb24-14"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() <span class="op">+</span></span>
<span id="cb24-15"><a href="#cb24-15"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span>(<span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">y =</span> contrastFit), <span class="dt">color =</span> <span class="st">'black'</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb24-16"><a href="#cb24-16"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_color_brewer</a></span>(<span class="st">"q-value"</span>, <span class="dt">palette =</span> <span class="st">"Spectral"</span>, <span class="dt">direction =</span> <span class="dv">-1</span>, <span class="dt">drop =</span> <span class="ot">FALSE</span>,</span>
<span id="cb24-17"><a href="#cb24-17"></a>                       <span class="dt">na.value =</span> <span class="st">"black"</span>) <span class="op">+</span></span>
<span id="cb24-18"><a href="#cb24-18"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span>(<span class="dt">yintercept =</span> <span class="dv">0</span>, <span class="dt">color =</span> <span class="st">'gray'</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb24-19"><a href="#cb24-19"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/guides.html">guides</a></span>(<span class="dt">color =</span> <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html">guide_legend</a></span>(<span class="dt">override.aes =</span> <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">alpha =</span> <span class="dv">1</span>))) <span class="op">+</span></span>
<span id="cb24-20"><a href="#cb24-20"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span>(. <span class="op">~</span><span class="st"> </span>cname) <span class="op">+</span></span>
<span id="cb24-21"><a href="#cb24-21"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span>(<span class="st">"HOXC9 8-mer differential specificity"</span>)</span></code></pre></div>
<p><img src="analysis-inference_files/figure-html/difspec-plot-1.png" width="960"></p>
<p>We see that unlike differential affinity, very few 8-mers are called in the differential specificity analysis for all alleles.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a>ds_dat <span class="op">%&gt;%</span></span>
<span id="cb25-2"><a href="#cb25-2"></a><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/tally.html">count</a></span>(cname, specificityQ) <span class="op">%&gt;%</span></span>
<span id="cb25-3"><a href="#cb25-3"></a><span class="st">    </span>tidyr<span class="op">::</span><span class="kw"><a href="https://tidyr.tidyverse.org/reference/spread.html">spread</a></span>(specificityQ, n, <span class="dt">fill =</span> <span class="dv">0</span>)</span></code></pre></div>
<pre><code>## # A tibble: 3 x 5
##   cname       `(0.01,1]` `(0.001,0.01]` `(1e-06,0.001]` `[0,1e-06]`
##   &lt;chr&gt;            &lt;dbl&gt;          &lt;dbl&gt;           &lt;dbl&gt;       &lt;dbl&gt;
## 1 HOXC9-K195R      32837             36              22           1
## 2 HOXC9-R193K      32896              0               0           0
## 3 HOXC9-R222W      32825             31              34           6</code></pre>
<p>We can similarly plot the residuals as well.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1"></a>ds_dat <span class="op">%&gt;%</span></span>
<span id="cb27-2"><a href="#cb27-2"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> specificityAxis)) <span class="op">+</span></span>
<span id="cb27-3"><a href="#cb27-3"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>(<span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">y =</span> contrastResidual, <span class="dt">color =</span> specificityQ), <span class="dt">alpha =</span> <span class="dv">1</span><span class="op">/</span><span class="dv">10</span>) <span class="op">+</span></span>
<span id="cb27-4"><a href="#cb27-4"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span>(<span class="dt">yintercept =</span> <span class="dv">0</span>, <span class="dt">color =</span> <span class="st">'gray'</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb27-5"><a href="#cb27-5"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() <span class="op">+</span></span>
<span id="cb27-6"><a href="#cb27-6"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_color_brewer</a></span>(<span class="st">"q-value"</span>, <span class="dt">palette =</span> <span class="st">"Spectral"</span>, <span class="dt">direction =</span> <span class="dv">-1</span>, <span class="dt">drop =</span> <span class="ot">FALSE</span>,</span>
<span id="cb27-7"><a href="#cb27-7"></a>                       <span class="dt">na.value =</span> <span class="st">"black"</span>) <span class="op">+</span></span>
<span id="cb27-8"><a href="#cb27-8"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/guides.html">guides</a></span>(<span class="dt">color =</span> <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html">guide_legend</a></span>(<span class="dt">override.aes =</span> <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">alpha =</span> <span class="dv">1</span>))) <span class="op">+</span></span>
<span id="cb27-9"><a href="#cb27-9"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span>(. <span class="op">~</span><span class="st"> </span>cname) <span class="op">+</span></span>
<span id="cb27-10"><a href="#cb27-10"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span>(<span class="st">"HOXC9 8-mer differential specificity (residuals)"</span>)</span></code></pre></div>
<p><img src="analysis-inference_files/figure-html/difspec-plot2-1.png" width="960"></p>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references hanging-indent">
<div id="ref-irizarry2003">
<p>Irizarry, Rafael A, Bridget Hobbs, Francois Collin, Yasmin D Beazer-Barclay, Kristen J Antonellis, Uwe Scherf, and Terence P Speed. 2003. “Exploration, Normalization, and Summaries of High Density Oligonucleotide Array Probe Level Data.” <em>Biostatistics</em> 4 (2): 249–64.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#hoxc9-dataset">HOXC9 Dataset</a></li>
      <li><a href="#sample-processing">Sample Processing</a></li>
      <li><a href="#preferential-affinity">Preferential Affinity</a></li>
      <li><a href="#differential-affinity">Differential Affinity</a></li>
      <li><a href="#differential-specificity">Differential Specificity</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://www.pkimes.com">Patrick Kimes</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
