<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>upbm: Summarization Details • upbm</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="upbm: Summarization Details">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">upbm</a>
        <span class="version label label-info" data-toggle="tooltip" data-placement="bottom" title="GitHub pre-release">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/upbm-quickstart.html">Quick Start</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">General</li>
    <li>
      <a href="../articles/upbm-quickstart.html">Quickstart Guide</a>
    </li>
    <li>
      <a href="../articles/upbm-classes.html">PBM Class Details</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Analysis</li>
    <li>
      <a href="../articles/analysis-preprocessing.html">Sample Pre-Processing</a>
    </li>
    <li>
      <a href="../articles/analysis-summarization.html">Probe and K-mer Summarization</a>
    </li>
    <li>
      <a href="../articles/analysis-inference.html">K-mer-Level Inference</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Other</li>
    <li>
      <a href="../articles/other-qcplots.html">Quality Control Plots</a>
    </li>
    <li>
      <a href="../articles/other-tidydata.html">Tidy PBM Data</a>
    </li>
  </ul>
</li>
<li>
  <a href="https://github.com/pkimes/upbm">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>upbm: Summarization Details</h1>
                        <h4 class="author">Patrick K. Kimes</h4>
            
            <h4 class="date">11 February 2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/pkimes/upbm/blob/master/vignettes/analysis-summarization.Rmd"><code>vignettes/analysis-summarization.Rmd</code></a></small>
      <div class="hidden name"><code>analysis-summarization.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>For an introduction to the <em>upbm</em> package, please see the quick start vignette (<code><a href="../articles/upbm-quickstart.html">vignette("upbm-quickstart")</a></code>). Here, we provide details on the summarization procedures implemented in the package for converting probe-level intensities to approximate 8-mer-level affinity estimates. In this vignette, we assume basic knowledge of the <em>PBMExperiment</em> class (see <code><a href="../articles/upbm-classes.html">vignette("upbm-classes")</a></code>). The same example HOXC9 dataset from the <em>upbmData</em> package is used in this vignette to illustrate the various functions.</p>
<p>To perform inference, the <em>upbm</em> pipeline aggregates normalized probe-level intensities to the 8-mer level across samples. <strong>These procedures require replicate data.</strong> When replicate samples are not available for conditions of interest, simple K-mer summary statistics can be computed using the <code><a href="../reference/summarizeKmers.html">upbm::summarizeKmers</a></code> function. However, these simple summary statistics cannot be used with downstream inference procedures implemented in the package (see <code><a href="../articles/analysis-inference.html">vignette("analysis-inference")</a></code>) and should be treated as exploratory statistics.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">"upbm"</span>))</span></code></pre></div>
<p>Throughout, we will be using <code><a href="https://rdrr.io/pkg/broom/man/reexports.html">broom::tidy</a></code> to tidy data stored in <em>PBMExperiment</em> and <em>SummarizedExperiment</em> objects for interactive analysis and plotting (see <code><a href="../articles/other-tidydata.html">vignette("other-tidydata")</a></code>). We load the <em>dplyr</em> package to help with interactive analysis.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="kw"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">"dplyr"</span>))</span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="kw"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">"ggplot2"</span>))</span></code></pre></div>
</div>
<div id="hoxc9-dataset" class="section level1">
<h1 class="hasAnchor">
<a href="#hoxc9-dataset" class="anchor"></a>HOXC9 Dataset</h1>
<p>For details on the example HOXC9 dataset, see the quick start vignette in this package or the <em>upbmData</em> package vignette. Here, we will use both Alexa488 and Cy3 scans.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(hoxc9alexa, <span class="dt">package =</span> <span class="st">"upbmData"</span>)</span>
<span id="cb3-2"><a href="#cb3-2"></a><span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(hoxc9cy3, <span class="dt">package =</span> <span class="st">"upbmData"</span>)</span></code></pre></div>
<p>Here, we again use just single PMT gain scan for the Alexa488 scans.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>alexa_subset &lt;-<span class="st"> </span>hoxc9alexa[, <span class="kw">colData</span>(hoxc9alexa)<span class="op">$</span>pmt <span class="op">==</span><span class="st"> </span><span class="dv">450</span>]</span></code></pre></div>
<p>We also load the reference Cy3 dataset.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(refcy3_8x60k_v1, <span class="dt">package =</span> <span class="st">"upbmAux"</span>)</span></code></pre></div>
</div>
<div id="sample-pre-processing" class="section level1">
<h1 class="hasAnchor">
<a href="#sample-pre-processing" class="anchor"></a>Sample Pre-Processing</h1>
<p>We will perform the earlier steps of pre-processing using the <code>upbm::pmbPreprocess</code> wrapper.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a>alexa_subset &lt;-<span class="st"> </span><span class="kw"><a href="../reference/upbmPreprocess.html">upbmPreprocess</a></span>(<span class="dt">pe =</span> alexa_subset,</span>
<span id="cb6-2"><a href="#cb6-2"></a>                               <span class="dt">cy3pe =</span> hoxc9cy3,</span>
<span id="cb6-3"><a href="#cb6-3"></a>                               <span class="dt">cy3refpe =</span> refcy3_8x60k_v1,</span>
<span id="cb6-4"><a href="#cb6-4"></a>                               <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</span></code></pre></div>
</div>
<div id="probe-aggregation" class="section level1">
<h1 class="hasAnchor">
<a href="#probe-aggregation" class="anchor"></a>Probe Aggregation</h1>
<p>The first step of summarization is to aggregate across replicate samples in the dataset. This is performed using the <code><a href="../reference/probeFit.html">upbm::probeFit</a></code> function. Given a set of samples, the function returned cross-replicate estimate probe-level affinity summaries as well as variance and degrees of freedom estimates.</p>
<p>As described in the quick start vignette, we must specify the column in the <em>colData</em> of the <em>PBMExperiment</em> containing replicate information. In our example dataset, this information in stored in the <code>"condition"</code> column.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="kw"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="kw">colData</span>(alexa_subset)<span class="op">$</span>condition)</span></code></pre></div>
<pre><code>## 
## HOXC9-K195R HOXC9-R193K HOXC9-R222W   HOXC9-REF 
##           2           2           3           3</code></pre>
<p>For each allelic variants, we only have 2 or 3 replicates. In PBM experiments, we often have access to only a few replicates of each condition. To obtain cross-replicate affinity and variance estimates, the <code><a href="../reference/probeFit.html">upbm::probeFit</a></code> internally calls <em>limma</em>, a popular package for the analysis of microarray data with features for handling analyses with low numbers of samples <span class="citation">(Ritchie et al. 2015)</span>.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>alexa_pfit &lt;-<span class="st"> </span><span class="kw"><a href="../reference/probeFit.html">probeFit</a></span>(alexa_subset, <span class="dt">stratify =</span> <span class="st">"condition"</span>)</span>
<span id="cb9-2"><a href="#cb9-2"></a>alexa_pfit</span></code></pre></div>
<pre><code>## class: PBMExperiment 
## dim: 41944 4 
## metadata(0):
## assays(3): beta sd df
## rownames: NULL
## rowData names(4): Column Row probeID Sequence
## colnames(4): HOXC9-K195R HOXC9-R193K HOXC9-R222W HOXC9-REF
## colData names(0):
## probeCols(4): Column Row probeID Sequence
## probeFilter names(0): 
## probeTrim: 1 36</code></pre>
<p>The returned <em>PBMExperiment</em> now only includes one column for each condition (<code>HOXC9-K195R</code>, <code>HOXC9-R193K</code>, <code>HOXC0-R222W</code>, <code>HOXC9-REF</code>). The cross-replicate summarized probe intensities are included in the <code>beta</code> assay.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a>pfit_dat &lt;-<span class="st"> </span>broom<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/broom/man/reexports.html">tidy</a></span>(alexa_pfit, <span class="dt">assay =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"beta"</span>, <span class="st">"sd"</span>))</span>
<span id="cb11-2"><a href="#cb11-2"></a>pfit_dat</span></code></pre></div>
<pre><code>## # A tibble: 167,776 x 7
##    Column   Row probeID      Sequence                  cname     beta    sd
##     &lt;int&gt; &lt;int&gt; &lt;chr&gt;        &lt;chr&gt;                     &lt;chr&gt;    &lt;dbl&gt; &lt;dbl&gt;
##  1      6     1 dBr_14334_J… GGTGTGAGTCCATTTCGTCAAACC… HOXC9-K…  11.6 0.482
##  2      7     1 dBr_06208_J… CAGTCTAAGTTTTCGGATTACCAT… HOXC9-K…  12.7 0.413
##  3      8     1 dBr_39317_J… CTTTTTAAAGACCTAGGAATCATT… HOXC9-K…  12.7 0.339
##  4      9     1 dBr_06987_J… CAGCTACGGAAGAGTAAAGGTTGT… HOXC9-K…  10.1 0.303
##  5     10     1 dBr_05074_J… GCTTCGAACGTGGGCGATGGAACA… HOXC9-K…  10.0 0.285
##  6     15     1 dBr_16182_J… CGCCCGTGGTAACAGTTACGAAAA… HOXC9-K…  12.7 0.485
##  7     23     1 dBr_20062_J… TTAGCCCGTTGTATGAGCAACTAT… HOXC9-K…  11.4 0.351
##  8     24     1 dBr_25542_J… TGCACAAGCGATCGAGCTCCTGTG… HOXC9-K…  10.5 0.369
##  9     26     1 dBr_40621_J… GGATGCCTCTTCTTTTCCGGGTCA… HOXC9-K…  10.7 0.434
## 10     27     1 dBr_21815_J… GTCAGAACCTCGCAAGTGGCACTT… HOXC9-K…  10.5 0.403
## # … with 167,766 more rows</code></pre>
<p>While raw probe intensities ranged from 0 to 2^16, the cross-sample summaries are computed after log2 transforming intensities. As a result, the <code>beta</code> values are mostly bounded between 0 and 16. The log2 transformation is performed with a pseudocount of 1 added to probe intensities to prevent <code>NA</code>s for zero intensity probes. This can be changed by modifying the <code>offset=</code> parameter in the call.</p>
<p>We can plot the distribution of <code>beta</code> estimates across conditions.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(pfit_dat, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> beta, <span class="dt">color =</span> cname)) <span class="op">+</span></span>
<span id="cb13-2"><a href="#cb13-2"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span>() <span class="op">+</span></span>
<span id="cb13-3"><a href="#cb13-3"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_color_brewer</a></span>(<span class="st">"condition"</span>, <span class="dt">palette =</span> <span class="st">"Set1"</span>) <span class="op">+</span></span>
<span id="cb13-4"><a href="#cb13-4"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() <span class="op">+</span></span>
<span id="cb13-5"><a href="#cb13-5"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span>(<span class="st">"HOXC9 cross-replicate summarized probe intensities"</span>)</span></code></pre></div>
<p><img src="analysis-summarization_files/figure-html/pfit-plot-1.png" width="480"></p>
<p>We see that most alleles have similar distributions with the exception of the R222W variant.</p>
</div>
<div id="k-mer-aggregation" class="section level1">
<h1 class="hasAnchor">
<a href="#k-mer-aggregation" class="anchor"></a>K-mer Aggregation</h1>
<p>Next, we take the cross-replicate aggregated probe intensities and further aggregate them to the 8-mer level using the <code><a href="../reference/kmerFit.html">upbm::kmerFit</a></code> function. The function takes the output from <code><a href="../reference/probeFit.html">upbm::probeFit</a></code> above along with a list of k-mers, usually all 8-mers, as the primary input. Mathematically, this is accomplished by framing the analysis as a meta-analysis across probes for each K-mer. K-mer-level estimates are computed by pooling the probe-level estimates for all probes containing the K-mer using the two-step DerSimonian and Laird estimator <span class="citation">(DerSimonian and Kacker 2007)</span>. Additional steps are also performed to adjust for position bias of K-mers along the probe sequence and to filter out outlier probes from each K-mer probe set. More details on this procedure can be found in the forthcoming manuscript describing the method.</p>
<p>To generate the full set of 8-mers, the package includes a helper function, <code><a href="../reference/uniqueKmers.html">upbm::uniqueKmers</a></code>. The function can be used to generate the complete set of K-mers of a specified length, unique to reverse complementation. We demonstrate this function by returning all unique 3-mers.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a><span class="kw"><a href="../reference/uniqueKmers.html">uniqueKmers</a></span>(<span class="dv">3</span>)</span></code></pre></div>
<pre><code>##  [1] "AAA" "AAC" "AAG" "AAT" "ACA" "ACC" "ACG" "ACT" "AGA" "AGC" "AGG"
## [12] "ATA" "ATC" "ATG" "CAA" "CAC" "CAG" "CCA" "CCC" "CCG" "CGA" "CGC"
## [23] "CTA" "CTC" "GAA" "GAC" "GCA" "GCC" "GGA" "GTA" "TAA" "TCA"</code></pre>
<p>As stated in the quick start, the function additionally computes statistics needed for downstream differential testing. More specifically, unless <code>contrasts = FALSE</code> is specified, the covariance between the baseline condition (column) and all other conditions is estimated during this step.</p>
<p>By default, if a single column is found containing the <code>REF</code> suffix, this will be used as the baseline condition. If no column includes the <code>REF</code> suffix or a different condition should be used as the baseline condition for computing contrasts, this should be specified to the <code>baseline=</code> parameter. In this analysis, we use the standard <code>HOXC9-REF</code> (wild type) allele as our baseline condition.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a>alexa_kfit &lt;-<span class="st"> </span><span class="kw"><a href="../reference/kmerFit.html">kmerFit</a></span>(alexa_pfit, <span class="dt">kmers =</span> <span class="kw"><a href="../reference/uniqueKmers.html">uniqueKmers</a></span>(8L),</span>
<span id="cb16-2"><a href="#cb16-2"></a>                      <span class="dt">baseline =</span> <span class="st">"HOXC9-REF"</span>)</span>
<span id="cb16-3"><a href="#cb16-3"></a>alexa_kfit</span></code></pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 32896 4 
## metadata(1): baseline
## assays(5): affinityEstimate affinityVariance contrastDifference
##   contrastAverage contrastVariance
## rownames: NULL
## rowData names(1): seq
## colnames(4): HOXC9-K195R HOXC9-R193K HOXC9-R222W HOXC9-REF
## colData names(0):</code></pre>
<p>The returned <em>PBMExperiment</em> includes the same number of columns corresponding to the four conditions. The number of rows is equal to the number of K-mers specified to <code>kmers=</code> above.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a><span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="kw"><a href="../reference/uniqueKmers.html">uniqueKmers</a></span>(8L))</span></code></pre></div>
<pre><code>## [1] 32896</code></pre>
<p>For each condition, an 8-mer <code>affinityEstimate</code> and <code>affinityVariance</code> is computed. For all conditions other than the baseline, <code>contrastDifference</code> (<code>baseline - condition</code>), <code>contrastAverage</code> (<code>(baseline + condition)/2</code>) and <code>contrastVariance</code> (<code><a href="https://rdrr.io/r/stats/cor.html">var(baseline - condition)</a></code>) values are also computed.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a>kfit_dat &lt;-<span class="st"> </span>broom<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/broom/man/reexports.html">tidy</a></span>(alexa_kfit, <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"affinityEstimate"</span>, <span class="st">"affinityVariance"</span>,</span>
<span id="cb20-2"><a href="#cb20-2"></a>                                      <span class="st">"contrastDifference"</span>, <span class="st">"contrastAverage"</span>,</span>
<span id="cb20-3"><a href="#cb20-3"></a>                                      <span class="st">"contrastVariance"</span>))</span>
<span id="cb20-4"><a href="#cb20-4"></a>kfit_dat</span></code></pre></div>
<pre><code>## # A tibble: 131,584 x 7
##    seq   cname affinityEstimate affinityVariance contrastDiffere…
##    &lt;chr&gt; &lt;chr&gt;            &lt;dbl&gt;            &lt;dbl&gt;            &lt;dbl&gt;
##  1 AAAA… HOXC…             11.8           0.0146          0.201  
##  2 AAAA… HOXC…             11.7           0.0156          0.111  
##  3 AAAA… HOXC…             11.5           0.0237          0.140  
##  4 AAAA… HOXC…             12.1           0.0225          0.275  
##  5 AAAA… HOXC…             11.7           0.0367          0.108  
##  6 AAAA… HOXC…             11.5           0.0222          0.0517 
##  7 AAAA… HOXC…             11.8           0.0239          0.00936
##  8 AAAA… HOXC…             11.4           0.0179          0.173  
##  9 AAAA… HOXC…             11.7           0.0308          0.0942 
## 10 AAAA… HOXC…             11.3           0.0234          0.0948 
## # … with 131,574 more rows, and 2 more variables: contrastAverage &lt;dbl&gt;,
## #   contrastVariance &lt;dbl&gt;</code></pre>
<p>All the values of the <code>contrast</code> columns are <code>NA</code> for the <code>HOXC9-REF</code> condition.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a>kfit_dat <span class="op">%&gt;%</span></span>
<span id="cb22-2"><a href="#cb22-2"></a><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(cname, <span class="kw"><a href="https://dplyr.tidyverse.org/reference/reexports.html">starts_with</a></span>(<span class="st">"contrast"</span>)) <span class="op">%&gt;%</span></span>
<span id="cb22-3"><a href="#cb22-3"></a><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(cname) <span class="op">%&gt;%</span></span>
<span id="cb22-4"><a href="#cb22-4"></a><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/summarise_all.html">summarize_if</a></span>(is.numeric, <span class="op">~</span><span class="kw"><a href="https://rdrr.io/r/base/mean.html">mean</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(.)))</span></code></pre></div>
<pre><code>## # A tibble: 4 x 4
##   cname       contrastDifference contrastAverage contrastVariance
##   &lt;chr&gt;                    &lt;dbl&gt;           &lt;dbl&gt;            &lt;dbl&gt;
## 1 HOXC9-K195R                  0               0                0
## 2 HOXC9-R193K                  0               0                0
## 3 HOXC9-R222W                  0               0                0
## 4 HOXC9-REF                    1               1                1</code></pre>
<p>We can plot the 8-mer affinity distributions across conditions.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(kfit_dat, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> affinityEstimate, <span class="dt">color =</span> cname)) <span class="op">+</span></span>
<span id="cb24-2"><a href="#cb24-2"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span>() <span class="op">+</span></span>
<span id="cb24-3"><a href="#cb24-3"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_color_brewer</a></span>(<span class="st">"condition"</span>, <span class="dt">palette =</span> <span class="st">"Set1"</span>) <span class="op">+</span></span>
<span id="cb24-4"><a href="#cb24-4"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() <span class="op">+</span></span>
<span id="cb24-5"><a href="#cb24-5"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span>(<span class="st">"HOXC9 cross-replicate summarized 8-mer affinities"</span>)</span></code></pre></div>
<p><img src="analysis-summarization_files/figure-html/kfit-plot-1.png" width="480"></p>
<p>Again, affinities appear similar for all alleles with the exception of the R222W variant.</p>
<p>We can also create MA plots using the <code>contrastDifference</code> and <code>contrastAverage</code> columns to compare each allelic variant against the baseline wild type condition. When generating this plot, we need to remember to filter out the <code>NA</code>s corresponding to the baseline condition’s values in the table.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a>kfit_dat <span class="op">%&gt;%</span></span>
<span id="cb25-2"><a href="#cb25-2"></a><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(contrastDifference)) <span class="op">%&gt;%</span></span>
<span id="cb25-3"><a href="#cb25-3"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> contrastAverage, <span class="dt">y =</span> contrastDifference)) <span class="op">+</span></span>
<span id="cb25-4"><a href="#cb25-4"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>(<span class="dt">alpha =</span> <span class="dv">1</span><span class="op">/</span><span class="dv">5</span>) <span class="op">+</span></span>
<span id="cb25-5"><a href="#cb25-5"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span>(. <span class="op">~</span><span class="st"> </span>cname) <span class="op">+</span><span class="st"> </span></span>
<span id="cb25-6"><a href="#cb25-6"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() <span class="op">+</span></span>
<span id="cb25-7"><a href="#cb25-7"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span>(<span class="st">"HOXC9 8-mer affinity MA plot"</span>)</span></code></pre></div>
<p><img src="analysis-summarization_files/figure-html/kfit-plot-ma-1.png" width="864"></p>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references hanging-indent">
<div id="ref-dersimonian2007">
<p>DerSimonian, Rebecca, and Raghu Kacker. 2007. “Random-Effects Model for Meta-Analysis of Clinical Trials: An Update.” <em>Contemporary Clinical Trials</em> 28 (2): 105–14.</p>
</div>
<div id="ref-ritchie2015">
<p>Ritchie, Matthew E, Belinda Phipson, Di Wu, Yifang Hu, Charity W Law, Wei Shi, and Gordon K Smyth. 2015. “Limma Powers Differential Expression Analyses for Rna-Sequencing and Microarray Studies.” <em>Nucleic Acids Research</em> 43 (7): e47–e47.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#hoxc9-dataset">HOXC9 Dataset</a></li>
      <li><a href="#sample-pre-processing">Sample Pre-Processing</a></li>
      <li><a href="#probe-aggregation">Probe Aggregation</a></li>
      <li><a href="#k-mer-aggregation">K-mer Aggregation</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://www.pkimes.com">Patrick Kimes</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
