<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>upbm: Analysis of universal Protein Binding Microarrays in R • upbm</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="upbm: Analysis of universal Protein Binding Microarrays in R">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">upbm</a>
        <span class="version label label-info" data-toggle="tooltip" data-placement="bottom" title="GitHub pre-release">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/upbm-quickstart.html">Quick Start</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">General</li>
    <li>
      <a href="../articles/upbm-quickstart.html">Quickstart Guide</a>
    </li>
    <li>
      <a href="../articles/upbm-classes.html">PBM Class Details</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Analysis</li>
    <li>
      <a href="../articles/analysis-preprocessing.html">Sample Pre-Processing</a>
    </li>
    <li>
      <a href="../articles/analysis-summarization.html">Probe and K-mer Summarization</a>
    </li>
    <li>
      <a href="../articles/analysis-inference.html">K-mer-Level Inference</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Other</li>
    <li>
      <a href="../articles/other-qcplots.html">Quality Control Plots</a>
    </li>
    <li>
      <a href="../articles/other-tidydata.html">Tidy PBM Data</a>
    </li>
  </ul>
</li>
<li>
  <a href="https://github.com/pkimes/upbm">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>upbm: Analysis of universal Protein Binding Microarrays in R</h1>
                        <h4 class="author">Patrick K. Kimes</h4>
            
            <h4 class="date">15 February 2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/pkimes/upbm/blob/master/vignettes/upbm-quickstart.Rmd"><code>vignettes/upbm-quickstart.Rmd</code></a></small>
      <div class="hidden name"><code>upbm-quickstart.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>“Protein binding microarrays (PBMs) allow for the high-throughput quantification of DNA binding specificities of proteins, such as transcription factors, in vitro. While several probe designs have been proposed for PBMs, universal PBMs (uPBMs) designed to cover all 10-mer DNA sequences, have enjoyed the most wide-spread adoption. The <em>upbm</em> package provides tools and methods for organizing probe-level PBM data from raw GPR scan files, summarizing PBM data at K-mer resolution, and performing statistical inference when replicate experiments are available. The package was developed for the analysis of uPBMs and may not be appropriate for other PBM designs. upbm package version: 0.99.0”</p>
    </div>
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>While a large collection of methods and software packages have been developed for processing and analyzing traditional DNA microarrays, the development of tools for protein binding microarrays (PBMs) has been limited. PBMs and DNA microarrays measure fundamentally different things, with PBMs detecting protein-DNA interactions and DNA microarrays assaying nucleic acid hybridization. Blindly applying DNA microarray tools to PBM data can lead to problems.</p>
<p>This package provides tools and methods for reading, processing, and analyzing PBM data using standard <a href="https://www.bioconductor.org/"><em>Bioconductor</em></a> <a href="http://bioconductor.org/developers/how-to/commonMethodsAndClasses/">data structures</a>. The primary focus is on universal-design PBMs (uPBMs) which provide approximately uniform coverage of all 8-mer DNA sequences across probes on the array <span class="citation">(Berger et al. 2006)</span>. <strong>The methods in this package have not been tested with other array designs.</strong></p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">"upbm"</span>))</span></code></pre></div>
<p>In this quickstart guide, we demonstrate the basic functionality of the <em>upbm</em> package. More details on the individual steps for pre-processing and inference, as well as the specialized classes (<em>PBMExperiment</em> and <em>PBMDesign</em>) are provided in other vignettes (<code><a href="../articles/upbm-classes.html">vignette("upbm-classes")</a></code>).</p>
</div>
<div id="hoxc9-dataset" class="section level1">
<h1 class="hasAnchor">
<a href="#hoxc9-dataset" class="anchor"></a>HOXC9 Dataset</h1>
<p>The input to the <em>upbm</em> package is raw GenePix Results (GPR) data for Alexa488 and Cy3 scans. Functions for reading GPR data into R are included with this package (see <code><a href="../articles/upbm-classes.html">vignette("upbm-classes")</a></code>). For this quickstart guide, we will use just the Alexa488 scans for a PBM dataset included in the <em>upbmData</em> package. Details on how Cy3 scans can also be used provided in a separate vignette (see <code>vignette("upbm-preprocessing")</code>).</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(hoxc9alexa, <span class="dt">package =</span> <span class="st">"upbmData"</span>)</span></code></pre></div>
<p>The dataset includes PBM data for four allelic variants of the HOXC9 human TF (wild type, K195R, R193K, R222W). For simplicity, we will subset this data to scans at a single PMT gain (450) and keep just the wild type and R222W alleles.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>alexa_subset &lt;-<span class="st"> </span>hoxc9alexa[, <span class="kw">colData</span>(hoxc9alexa)<span class="op">$</span>pmt <span class="op">==</span><span class="st"> </span><span class="dv">450</span>]</span>
<span id="cb3-2"><a href="#cb3-2"></a>alexa_subset &lt;-<span class="st"> </span>alexa_subset[, <span class="kw">colData</span>(alexa_subset)<span class="op">$</span>condition <span class="op">%in%</span></span>
<span id="cb3-3"><a href="#cb3-3"></a><span class="st">                                                    </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"HOXC9-REF"</span>, <span class="st">"HOXC9-R222W"</span>)]</span></code></pre></div>
<p>The data is stored as a <em>PBMExperiment</em> object, which extends the standard Bioconductor <em>SummarizedExperiment</em> class. Each sample is stored as a column, with rows corresponding to probes on the array.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>alexa_subset</span></code></pre></div>
<pre><code>## class: PBMExperiment 
## dim: 62976 6 
## metadata(0):
## assays(2): fore back
## rownames: NULL
## rowData names(4): Column Row probeID Sequence
## colnames(6): s4 s7 ... s33 s34
## colData names(10): date version ... condition id_idx
## probeCols(4): Column Row probeID Sequence
## probeFilter names(1): probeID
## probeTrim: 1 36</code></pre>
</div>
<div id="sample-pre-processing" class="section level1">
<h1 class="hasAnchor">
<a href="#sample-pre-processing" class="anchor"></a>Sample Pre-Processing</h1>
<p>Analysis of uPBM data is performed in three steps: sample pre-processing, 8-mer summarization, and 8-mer inference. First, samples are preprocessed using <code><a href="../reference/upbmPreprocess.html">upbm::upbmPreprocess</a></code>. This wrapper functions performs various pre-processing steps, including Cy3 normalization. In this quickstart, we will not be using Cy3 scans and instead set the <code>cy3pe=</code> option to <code>NULL</code>. We also silence verbose output here.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a>alexa_normed &lt;-<span class="st"> </span><span class="kw"><a href="../reference/upbmPreprocess.html">upbmPreprocess</a></span>(<span class="dt">pe =</span> alexa_subset, <span class="dt">cy3pe =</span> <span class="ot">NULL</span>,</span>
<span id="cb6-2"><a href="#cb6-2"></a>                               <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</span></code></pre></div>
<p>The returned object is a <em>PBMExperiment</em> object with the same number of columns as the original <em>PBMExperiment</em>, with normalized probe intensities now stored in the <code>normalized</code> assay. Rows have been filtered to remove background and control probes.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>alexa_normed</span></code></pre></div>
<pre><code>## class: PBMExperiment 
## dim: 41944 6 
## metadata(0):
## assays(4): normalized fore back spatialbias
## rownames: NULL
## rowData names(4): Column Row probeID Sequence
## colnames(6): s4 s7 ... s33 s34
## colData names(13): date version ... acrossRepMultScale
##   acrossRepAddScale
## probeCols(4): Column Row probeID Sequence
## probeFilter names(0): 
## probeTrim: 1 36</code></pre>
</div>
<div id="mer-summarization" class="section level1">
<h1 class="hasAnchor">
<a href="#mer-summarization" class="anchor"></a>8-mer Summarization</h1>
<p>After normalization, probe-level intensities are aggregated across replicates by fitting a probe-level model to the data. Since the dataset includes samples for different conditions, this must be specified. In our dataset, this information is included in the <code>"condition"</code> column of the <code>colData</code> of the <em>PBMExperiment</em>. This is passed to the <code>stratify=</code> parameter.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>alexa_pfit &lt;-<span class="st"> </span><span class="kw"><a href="../reference/probeFit.html">probeFit</a></span>(alexa_normed, <span class="dt">stratify =</span> <span class="st">"condition"</span>)</span></code></pre></div>
<p>The resulting object is a <em>PBMExperiment</em> with the same number of rows, but only one column for each unique value in the column specified by <code>stratify=</code>. In this case, we have two conditions, <code>HOXC9-REF</code> and <code>HOXC9-R222W</code>. Aggregated probe summaries are stored in the <code>beta</code> assay.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a>alexa_pfit</span></code></pre></div>
<pre><code>## class: PBMExperiment 
## dim: 41944 2 
## metadata(0):
## assays(3): beta sd df
## rownames: NULL
## rowData names(4): Column Row probeID Sequence
## colnames(2): HOXC9-R222W HOXC9-REF
## colData names(0):
## probeCols(4): Column Row probeID Sequence
## probeFilter names(0): 
## probeTrim: 1 36</code></pre>
<p>Next, probe-level summaries are used to obtain 8-mer-level summaries. A helper function, <code><a href="../reference/uniqueKmers.html">upbm::uniqueKmers</a></code> is provided for generating the full set of K-mers of a specified length (unique to reverse complementation). During this step, parameters are also estimated for downstream differential testing and inference. The baseline condition must be specified to the <code>baseline=</code> parameter.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a>alexa_kfit &lt;-<span class="st"> </span><span class="kw"><a href="../reference/kmerFit.html">kmerFit</a></span>(alexa_pfit, <span class="dt">kmers =</span> <span class="kw"><a href="../reference/uniqueKmers.html">uniqueKmers</a></span>(8L),</span>
<span id="cb12-2"><a href="#cb12-2"></a>                      <span class="dt">baseline =</span> <span class="st">"HOXC9-REF"</span>)</span></code></pre></div>
<p>The result is a <em>SummarizedExperiment</em> with K-mer-level affinity and variance estimates for each condition, with rows now corresponding to K-mers rather than probes.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a>alexa_kfit</span></code></pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 32896 2 
## metadata(1): baseline
## assays(5): affinityEstimate affinityVariance contrastDifference
##   contrastAverage contrastVariance
## rownames: NULL
## rowData names(1): seq
## colnames(2): HOXC9-R222W HOXC9-REF
## colData names(0):</code></pre>
</div>
<div id="mer-inference" class="section level1">
<h1 class="hasAnchor">
<a href="#mer-inference" class="anchor"></a>8-mer Inference</h1>
<p>To test for differential 8-mer affinity across conditions between a baseline condition (e.g. the wild type HOXC9 TF) and all other conditions, we make a call to <code><a href="../reference/kmerTestContrast.html">upbm::kmerTestContrast</a></code>. Other methods for inference are also defined in <code><a href="../reference/kmerTestAffinity.html">upbm::kmerTestAffinity</a></code> and <code><a href="../reference/kmerTestSpecificity.html">upbm::kmerTestSpecificity</a></code>. Note that all differential affinity and specificity calls are relative to the baseline condition specified above.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a>alexa_kdiff &lt;-<span class="st"> </span><span class="kw"><a href="../reference/kmerTestContrast.html">kmerTestContrast</a></span>(alexa_kfit)</span></code></pre></div>
<p>The resulting object is again a <em>SummarizedExperiment</em> with assays containing the results of differential affinity testing.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a>alexa_kdiff</span></code></pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 32896 2 
## metadata(0):
## assays(6): contrastAverage contrastDifference ... contrastP
##   contrastQ
## rownames: NULL
## rowData names(1): seq
## colnames(2): HOXC9-R222W HOXC9-REF
## colData names(0):</code></pre>
<p>K-mer level averages and differences between the baseline and all other conditions are included in the <code>contrastAverage</code> and <code>contrastDifference</code> assays. FDR-adjusted p-values are included in the <code>contrastQ</code> assay. We can tidy the results from a <em>PBMExperiment</em> to a <em>tibble</em> using the <code><a href="https://rdrr.io/pkg/broom/man/reexports.html">broom::tidy</a></code> function.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a>diffdat &lt;-<span class="st"> </span>broom<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/broom/man/reexports.html">tidy</a></span>(alexa_kdiff, <span class="dt">assay =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"contrastAverage"</span>,</span>
<span id="cb18-2"><a href="#cb18-2"></a>                                              <span class="st">"contrastDifference"</span>,</span>
<span id="cb18-3"><a href="#cb18-3"></a>                                              <span class="st">"contrastQ"</span>))</span>
<span id="cb18-4"><a href="#cb18-4"></a>diffdat &lt;-<span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(diffdat, <span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(contrastAverage))</span>
<span id="cb18-5"><a href="#cb18-5"></a>diffdat</span></code></pre></div>
<pre><code>## # A tibble: 32,896 x 5
##    seq      cname       contrastAverage contrastDifference  contrastQ
##    &lt;chr&gt;    &lt;chr&gt;                 &lt;dbl&gt;              &lt;dbl&gt;      &lt;dbl&gt;
##  1 AAAAAAAA HOXC9-R222W            11.5            -0.340  0.00485   
##  2 AAAAAAAC HOXC9-R222W            11.4            -0.405  0.00322   
##  3 AAAAAAAG HOXC9-R222W            11.3            -0.0669 0.578     
##  4 AAAAAAAT HOXC9-R222W            11.6            -0.418  0.000297  
##  5 AAAAAACA HOXC9-R222W            11.4            -0.350  0.0232    
##  6 AAAAAACC HOXC9-R222W            11.3            -0.257  0.0791    
##  7 AAAAAACG HOXC9-R222W            11.5            -0.559  0.00000616
##  8 AAAAAACT HOXC9-R222W            11.2            -0.0980 0.391     
##  9 AAAAAAGA HOXC9-R222W            11.4            -0.412  0.00108   
## 10 AAAAAAGC HOXC9-R222W            11.2            -0.0273 0.853     
## # … with 32,886 more rows</code></pre>
<p>Assays are now contained in columns with rows corresponding to single samples and rows from the <em>PBMExperiment</em>. The <code>tidy</code> function can be used with any of the <em>PBMExperiment</em> objects above. More details are provided in <code><a href="../articles/other-tidydata.html">vignette("other-tidydata")</a></code>. Using the tidy data, plot the results of differential testing is now relatively straightforward. We can create a standard MA plot by plotting the <code>contrastAverage</code> and <code>contrastDifference</code> columns as the X and Y axes.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a><span class="kw"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">"ggplot2"</span>))</span>
<span id="cb20-2"><a href="#cb20-2"></a></span>
<span id="cb20-3"><a href="#cb20-3"></a><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(diffdat, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> contrastAverage,</span>
<span id="cb20-4"><a href="#cb20-4"></a>                    <span class="dt">y =</span> contrastDifference,</span>
<span id="cb20-5"><a href="#cb20-5"></a>                    <span class="dt">color =</span> contrastQ <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.001</span>)) <span class="op">+</span></span>
<span id="cb20-6"><a href="#cb20-6"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>(<span class="dt">alpha =</span> <span class="dv">1</span><span class="op">/</span><span class="dv">10</span>) <span class="op">+</span></span>
<span id="cb20-7"><a href="#cb20-7"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span>(<span class="dt">yintercept =</span> <span class="dv">0</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb20-8"><a href="#cb20-8"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_color_brewer</a></span>(<span class="dt">palette =</span> <span class="st">"Set1"</span>, <span class="dt">direction =</span> <span class="dv">-1</span>) <span class="op">+</span></span>
<span id="cb20-9"><a href="#cb20-9"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() <span class="op">+</span></span>
<span id="cb20-10"><a href="#cb20-10"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/guides.html">guides</a></span>(<span class="dt">color =</span> <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html">guide_legend</a></span>(<span class="dt">override.aes =</span> <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">alpha =</span> <span class="dv">1</span>))) <span class="op">+</span></span>
<span id="cb20-11"><a href="#cb20-11"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span>(. <span class="op">~</span><span class="st"> </span>cname) <span class="op">+</span></span>
<span id="cb20-12"><a href="#cb20-12"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span>(<span class="st">"HOXC9 differential 8-mer affinity MA plot"</span>)</span></code></pre></div>
<p><img src="upbm-quickstart_files/figure-html/quick-plot-ma-1.png" width="576"></p>
<p>Additionally, we can also take a look at the distribution of 8-mer-level differential affinities between conditions.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(diffdat, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> contrastDifference,</span>
<span id="cb21-2"><a href="#cb21-2"></a>                    <span class="dt">fill =</span> contrastQ <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.001</span>)) <span class="op">+</span></span>
<span id="cb21-3"><a href="#cb21-3"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span>(<span class="dt">color =</span> <span class="st">'black'</span>, <span class="dt">binwidth =</span> <span class="fl">.05</span>, <span class="dt">boundary =</span> <span class="dv">0</span>) <span class="op">+</span></span>
<span id="cb21-4"><a href="#cb21-4"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span>(<span class="dt">xintercept =</span> <span class="dv">0</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb21-5"><a href="#cb21-5"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_fill_brewer</a></span>(<span class="dt">palette =</span> <span class="st">"Set1"</span>, <span class="dt">direction =</span> <span class="dv">-1</span>) <span class="op">+</span></span>
<span id="cb21-6"><a href="#cb21-6"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() <span class="op">+</span></span>
<span id="cb21-7"><a href="#cb21-7"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/guides.html">guides</a></span>(<span class="dt">color =</span> <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html">guide_legend</a></span>(<span class="dt">override.aes =</span> <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">alpha =</span> <span class="dv">1</span>))) <span class="op">+</span></span>
<span id="cb21-8"><a href="#cb21-8"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span>(. <span class="op">~</span><span class="st"> </span>cname) <span class="op">+</span></span>
<span id="cb21-9"><a href="#cb21-9"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span>(<span class="st">"HOXC9 differential 8-mer-level affinity"</span>)</span></code></pre></div>
<p><img src="upbm-quickstart_files/figure-html/quick-plot-dist-1.png" width="576"></p>
<p>Using the tabular data, we can also identify the set of 8-mers with largest differences between the two conditions.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a><span class="kw"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">"dplyr"</span>))</span>
<span id="cb22-2"><a href="#cb22-2"></a></span>
<span id="cb22-3"><a href="#cb22-3"></a>diffdat <span class="op">%&gt;%</span></span>
<span id="cb22-4"><a href="#cb22-4"></a><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/top_n.html">top_n</a></span>(<span class="dv">10</span>, <span class="kw"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(contrastDifference)) <span class="op">%&gt;%</span></span>
<span id="cb22-5"><a href="#cb22-5"></a><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span>(contrastDifference) <span class="op">%&gt;%</span></span>
<span id="cb22-6"><a href="#cb22-6"></a><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(cname, seq, contrastDifference, contrastQ)</span></code></pre></div>
<pre><code>## # A tibble: 10 x 4
##    cname       seq      contrastDifference contrastQ
##    &lt;chr&gt;       &lt;chr&gt;                 &lt;dbl&gt;     &lt;dbl&gt;
##  1 HOXC9-R222W GTCGTAAA              -2.36 1.74e-195
##  2 HOXC9-R222W GTCGTTAA              -2.23 5.70e-129
##  3 HOXC9-R222W GTCATTAA              -2.17 1.50e-197
##  4 HOXC9-R222W GTAATTAC              -2.15 1.87e-119
##  5 HOXC9-R222W GGTAATTA              -2.11 0.       
##  6 HOXC9-R222W GGTCGTTA              -2.05 1.30e- 90
##  7 HOXC9-R222W GCCATTAA              -2.04 6.42e-152
##  8 HOXC9-R222W CCATTAAA              -2.02 2.65e-116
##  9 HOXC9-R222W CCCATTAA              -2.01 9.13e-152
## 10 HOXC9-R222W ATCGTAAA              -2.00 2.06e-141</code></pre>
</div>
<div id="other-vignettes" class="section level1">
<h1 class="hasAnchor">
<a href="#other-vignettes" class="anchor"></a>Other Vignettes</h1>
<p>More details on various components of the package are provided in several other vignettes included with the package.</p>
<ul>
<li>
<code><a href="../articles/upbm-classes.html">vignette("upbm-classes")</a></code>: Details on the two uPBM classes defined in the package.</li>
<li>
<code><a href="../articles/analysis-preprocessing.html">vignette("analysis-preprocessing")</a></code>: Details on the pre-processing procedures.</li>
<li>
<code><a href="../articles/analysis-summarization.html">vignette("analysis-summarization")</a></code>: Details on the steps for summarizing probes to 8-mers.</li>
<li>
<code><a href="../articles/analysis-inference.html">vignette("analysis-inference")</a></code>: Details on the various K-mer-level inference procedures.</li>
<li>
<code><a href="../articles/other-qcplots.html">vignette("other-qcplots")</a></code>: Details on various QC plots useful when evaluating the quality of PBM data.</li>
<li>
<code><a href="../articles/other-tidydata.html">vignette("other-tidydata")</a></code>: Details on how data can be tidied from <em>PBMExperiment</em> to <em>tibble</em> format.</li>
</ul>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references hanging-indent">
<div id="ref-berger2006">
<p>Berger, Michael F, Anthony A Philippakis, Aaron M Qureshi, Fangxue S He, Preston W Estep III, and Martha L Bulyk. 2006. “Compact, Universal Dna Microarrays to Comprehensively Determine Transcription-Factor Binding Site Specificities.” <em>Nature Biotechnology</em> 24 (11): 1429–35.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#hoxc9-dataset">HOXC9 Dataset</a></li>
      <li><a href="#sample-pre-processing">Sample Pre-Processing</a></li>
      <li><a href="#mer-summarization">8-mer Summarization</a></li>
      <li><a href="#mer-inference">8-mer Inference</a></li>
      <li><a href="#other-vignettes">Other Vignettes</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://www.pkimes.com">Patrick Kimes</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
