<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>upbm: Analysis of Protein Binding Microarrays in R • upbm</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="upbm: Analysis of Protein Binding Microarrays in R">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">upbm</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/upbm.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/upbm-classes.html">upbm: Class Details</a>
    </li>
    <li>
      <a href="../articles/upbm-preprocessing.html">upbm: Pre-Processing Details</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/pkimes/upbm">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>upbm: Analysis of Protein Binding Microarrays in R</h1>
                        <h4 class="author">Patrick K. Kimes</h4>
            
            <h4 class="date">23 December 2019</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/pkimes/upbm/blob/master/vignettes/upbm.Rmd"><code>vignettes/upbm.Rmd</code></a></small>
      <div class="hidden name"><code>upbm.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>“Protein binding microarrays (PBMs) allow for the high-throughput quantification of DNA binding specificities of proteins, such as transcription factors, in vitro. While several probe designs have been proposed for PBMs, universal PBMs (uPBMs) designed to cover all 10-mer DNA sequences, have enjoyed the most wide-spread adoption. The <em>upbm</em> package provides tools and methods for organizing probe-level PBM data from raw GPR scan files, summarizing PBM data at K-mer resolution, and performing statistical inference when replicate experiments are available. The package was developed for the analysis of uPBMs, but should be appropriate for many other PBM designs as well. upbm package version: 0.99.0”</p>
    </div>
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>While a large collection of methods and software packages have been developed for the processing and analysis of traditional DNA microarrays, the development of tools for protein binding microarrays (PBMs) has been limited. Due to the substantial difference of what is being measured by PBMs (protein-DNA binding) and DNA microarrays (nucleic acid hybridization), blindly applying DNA microarray tools to PBM data can lead to unexpected problems.</p>
<p>This package provides tools and methods for reading, processing, and analyzing PBM data in <em>R</em> using standard <a href="https://www.bioconductor.org/"><em>Bioconductor</em></a> <a href="http://bioconductor.org/developers/how-to/commonMethodsAndClasses/">data structures</a>. The primary focus is on universal-design PBMs (uPBMs) which provide approximately uniform coverage of all 8-mer DNA sequences across probes on the array <span class="citation">(Berger et al. 2006)</span>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">"upbm"</span>))</span></code></pre></div>
<p>In this vignette we demonstrate the typical analysis steps with the <em>upbm</em> package. Here, and in other vignettes, data from a triplicate series of PBM experiments are used to illustrate the features of the package. While this vignette covers only the basics of the <em>upbm</em> package, more details on the object classes (<em>PBMExperiment</em> and <em>PBMDesign</em>) and pre-processing functions are also provided in separate vignettes.</p>
</div>
<div id="hoxc9-dataset" class="section level1">
<h1 class="hasAnchor">
<a href="#hoxc9-dataset" class="anchor"></a>HOXC9 Dataset</h1>
<p>The dataset includes GenePix Results (GPR) data from Alexa488 scans for wild type human transcription factor (TF) HOXC9 and three allelic variants (K195R, R193K, R222W). Corresponding Cy3 scans are also included for all arrays. Each array was scanned for Alexa488 fluorescence at multiple PMT gains and Cy3 at a single PMT gain. The series comes from a larger set of experiments studying the effect of allelic variants on human transcription factor binding (manuscript in preparation). The HOXC9 data objects are stored in a separate <em>upbmData</em> package for convenience and include a table of sample metadata (<code>hoxc9table</code>), a <em>PBMExperiment</em> of Alexa488 scan data (<code>hoxc9alexa</code>), and a <em>PBMExperiment</em> of Cy3 scan data (<code>hoxc9cy3</code>).</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(hoxc9table, <span class="dt">package =</span> <span class="st">"upbmData"</span>)</span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(hoxc9alexa, <span class="dt">package =</span> <span class="st">"upbmData"</span>)</span>
<span id="cb2-3"><a href="#cb2-3"></a><span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(hoxc9cy3, <span class="dt">package =</span> <span class="st">"upbmData"</span>)</span></code></pre></div>
<p>Metadata for Alexa488 and Cy3 scans are provided in the sample table.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>hoxc9table</span></code></pre></div>
<pre><code># # A tibble: 40 x 11
#    gpr     date version    id reuse type    pmt   idx target condition
#    &lt;chr&gt;  &lt;dbl&gt; &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;    
#  1 ./PB… 170606 v14       226     1 Alexa   400     2 HOXC9  HOXC9-REF
#  2 ./PB… 170606 v14       226     1 Alexa   400     3 HOXC9  HOXC9-R1…
#  3 ./PB… 170606 v14       226     1 Alexa   400     6 HOXC9  HOXC9-K1…
#  4 ./PB… 170606 v14       226     1 Alexa   450     2 HOXC9  HOXC9-REF
#  5 ./PB… 170606 v14       226     1 Alexa   450     3 HOXC9  HOXC9-R1…
#  6 ./PB… 170606 v14       226     1 Alexa   450     6 HOXC9  HOXC9-K1…
#  7 ./PB… 170606 v14       226     1 Alexa   450     7 HOXC9  HOXC9-R2…
#  8 ./PB… 170606 v14       226     1 Alexa   500     2 HOXC9  HOXC9-REF
#  9 ./PB… 170606 v14       226     1 Alexa   500     3 HOXC9  HOXC9-R1…
# 10 ./PB… 170606 v14       226     1 Alexa   500     6 HOXC9  HOXC9-K1…
# # … with 30 more rows, and 1 more variable: id_idx &lt;chr&gt;</code></pre>
<p>Several important metadata variables are recorded in the table - notably, <code>type</code> (scan type), <code>gpr</code> (path to the GPR file), <code>lp</code> (PMT gain of the scan), and <code>condition</code> (HOXC9 allele). Details on how the raw GPR files for the HOXC9 data can be accessed and loaded to create <code>hoxc9table</code>, <code>hoxc9alexa</code>, and <code>hoxc9cy3</code> are described in the <em>upbmData</em> vignette.</p>
<div id="illustrative-subset" class="section level2">
<h2 class="hasAnchor">
<a href="#illustrative-subset" class="anchor"></a>Illustrative Subset</h2>
<p>The HOXC9 example data includes multiple Alexa488 scans of each array at different PMT gains. Since we only require a single scan for each array, we pick a high PMT gain value for each replicate with no probe intensities reaching saturation (intensity = 2^16). In this case, since all samples were scanned at PMT gain 450, we will use these. Since probe log intensities scale linearly with PMT gain, any differences will be corrected during normalization steps, and the exact choice of PMT gain value is largely unimportant.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>alexa_subset &lt;-<span class="st"> </span>hoxc9alexa[, <span class="kw">colData</span>(hoxc9alexa)<span class="op">$</span>pmt <span class="op">==</span><span class="st"> </span><span class="dv">450</span>]</span></code></pre></div>
<p>In the interest of keeping the run time of this vignette manageable, we also reduce the dataset from the wild type and three allelic variants to just the wild type allele and a single variant allele (<code>HOXC9-R222W</code>).</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a>alexa_subset &lt;-<span class="st"> </span>alexa_subset[, <span class="kw">colData</span>(alexa_subset)<span class="op">$</span>condition <span class="op">%in%</span></span>
<span id="cb6-2"><a href="#cb6-2"></a><span class="st">                                                    </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"HOXC9-REF"</span>, <span class="st">"HOXC9-R222W"</span>)]</span>
<span id="cb6-3"><a href="#cb6-3"></a></span>
<span id="cb6-4"><a href="#cb6-4"></a>cy3_subset &lt;-<span class="st"> </span>hoxc9cy3[, <span class="kw">colData</span>(hoxc9cy3)<span class="op">$</span>id_idx <span class="op">%in%</span></span>
<span id="cb6-5"><a href="#cb6-5"></a><span class="st">                                          </span><span class="kw">colData</span>(alexa_subset)<span class="op">$</span>id_idx]</span></code></pre></div>
</div>
</div>
<div id="tidy-data" class="section level1">
<h1 class="hasAnchor">
<a href="#tidy-data" class="anchor"></a>Tidy Data</h1>
<p>“Tidy data” has become a popular and powerful framework for organizing data during interactive analysis. In the tidy data framework, data is organized as a data.frame with each row corresponding to an individual obervation or sample. Not only does the tidy data framework help keep data organized, but it also unlocks the powerful data parsing and visualization functions in the <a href="https://www.tidyverse.org/">Tidyverse</a> collection of packages.</p>
<p>To keep track of various probe and sample metadata compactly, uPBM data are not organized as tidy data. Instead, they are stored as <em>PBMExperiment</em> and <em>PBMDesign</em> objects which extend core Bioconductor data structures (see separate vignette: <em>upbm: Class Details</em>). However, when performing interactive analysis, it can be useful to extract tidy data from the <em>PBMExperiment</em> objects. This is made possible with the <code><a href="https://rdrr.io/pkg/broom/man/reexports.html">broom::tidy</a></code> function originally defined in the <a href="https://cran.r-project.org/web/packages/broom/"><em>broom</em></a> package.</p>
<p>The data for a single assay in <em>PBMExperiment</em> and <em>SummarizedExperiment</em> objects can be returned in wide or long format by passing the objects to <code><a href="https://rdrr.io/pkg/broom/man/reexports.html">broom::tidy</a></code>.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>broom<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/broom/man/reexports.html">tidy</a></span>(alexa_subset, <span class="dt">assay =</span> <span class="st">"fore"</span>, <span class="dt">long =</span> <span class="ot">FALSE</span>)</span></code></pre></div>
<pre><code>## # A tibble: 41,944 x 10
##       s4    s7   s17   s20   s33   s34 Column   Row probeID   Sequence     
##    &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;int&gt; &lt;int&gt; &lt;chr&gt;     &lt;chr&gt;        
##  1  9522  1137  2305  2478   766  2225      6     1 dBr_1433… GGTGTGAGTCCA…
##  2 18050  1209  4903  2662  1582  1442      7     1 dBr_0620… CAGTCTAAGTTT…
##  3 19591  1140  4365  2523  1539  2074      8     1 dBr_3931… CTTTTTAAAGAC…
##  4  4150   720  1029  1307   297  1068      9     1 dBr_0698… CAGCTACGGAAG…
##  5  4499   708   966  1583   342  1334     10     1 dBr_0507… GCTTCGAACGTG…
##  6 16744  1298  3956  2889   719  1593     15     1 dBr_1618… CGCCCGTGGTAA…
##  7  7032   812  2215  1564   542  1345     23     1 dBr_2006… TTAGCCCGTTGT…
##  8  3493   558  1411  1180   330   927     24     1 dBr_2554… TGCACAAGCGAT…
##  9  3231   725  1244  1968   271   761     26     1 dBr_4062… GGATGCCTCTTC…
## 10  4323   663  1157  1408   340   950     27     1 dBr_2181… GTCAGAACCTCG…
## # … with 41,934 more rows</code></pre>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>broom<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/broom/man/reexports.html">tidy</a></span>(alexa_subset, <span class="dt">assay =</span> <span class="st">"fore"</span>, <span class="dt">long =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<pre><code>## # A tibble: 251,664 x 16
##    Column   Row probeID Sequence cname value   date version    id reuse
##     &lt;int&gt; &lt;int&gt; &lt;chr&gt;   &lt;chr&gt;    &lt;chr&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt;
##  1      6     1 dBr_14… GGTGTGA… s4     9522 170606 v14       226     1
##  2      7     1 dBr_06… CAGTCTA… s4    18050 170606 v14       226     1
##  3      8     1 dBr_39… CTTTTTA… s4    19591 170606 v14       226     1
##  4      9     1 dBr_06… CAGCTAC… s4     4150 170606 v14       226     1
##  5     10     1 dBr_05… GCTTCGA… s4     4499 170606 v14       226     1
##  6     15     1 dBr_16… CGCCCGT… s4    16744 170606 v14       226     1
##  7     23     1 dBr_20… TTAGCCC… s4     7032 170606 v14       226     1
##  8     24     1 dBr_25… TGCACAA… s4     3493 170606 v14       226     1
##  9     26     1 dBr_40… GGATGCC… s4     3231 170606 v14       226     1
## 10     27     1 dBr_21… GTCAGAA… s4     4323 170606 v14       226     1
## # … with 251,654 more rows, and 6 more variables: type &lt;chr&gt;, pmt &lt;dbl&gt;,
## #   idx &lt;dbl&gt;, target &lt;chr&gt;, condition &lt;chr&gt;, id_idx &lt;chr&gt;</code></pre>
<p>Notice that when <code>long = FALSE</code>, the output contains only rowData and the values of the assay as a <em>tibble</em> with dimensions matching the input <em>PBMExperiment</em>. Here, each row corresponds to a single probe. When <code>long = TRUE</code>, a longer <em>tibble</em> is returned with each row corresponding to a single probe and single sample (column) of the input <em>PBMExperiment</em>. In the longer format, colData is also included in the output. When the object is a <em>PBMExperiment</em> object, only probes passing the filtering criteria specified in the <em>PBMDesign</em> object are included in the output (for more details, see: <em>upbm: Class Details</em>). In this case, this includes all probes with probe IDs prefixed by <code>dBr_</code>, corresponding to the subset of de Bruijn sequence probes on the array.</p>
</div>
<div id="quickstart" class="section level1">
<h1 class="hasAnchor">
<a href="#quickstart" class="anchor"></a>Quickstart</h1>
<p>Using the example data included in the <em>upbmData</em> package, we first quickly demonstrate how GPR data can be analyzed to obtain 8-mer-level affinity summaries and used to perform differential affinity analysis. Data analysis proceeds in the following steps:</p>
<ol style="list-style-type: decimal">
<li>load GPR data as <em>PBMExperiment</em> object,</li>
<li>perform data pre-processing, including normalization and filtering,</li>
<li>perform probe-level summarization across replicates,</li>
<li>perform K-mer-level summarization across probes,</li>
<li>perform K-mer-level inference.</li>
</ol>
<p>First, input data should be available as GenePix Results (GPR) files, with each GPR file corresponding to a single sample scan. Metadata for the GPR files, including paths to the files, must be organized in a data.frame with each row corresponding to a scan (see <code>hoxc9table</code> above). The path to the GPR file <strong>must</strong> be specified in a column named <code>gpr</code>. The data.frame should also include any relevant metadata about the scan and sample (e.g. scan parameters or properties of the assayed protein). Both Alexa488 and Cy3 scans for the example dataset in the <em>upbmData</em> package are provided in <code>hoxc9table</code>.</p>
<p>GPR files listed in the data.frame are read in using <code><a href="../reference/gpr2PBMExperiment.html">gpr2PBMExperiment(..)</a></code>. If possible, a corresponding probe design should also be specified. The probe design should be a <em>PBMDesign</em> object. <em>PBMDesign</em> objects for standard uPBM designs have been made available in the <em>upbmAux</em> package, including the standard 8x60k design used for this dataset.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a><span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(pbm_8x60k_v1, <span class="dt">package =</span> <span class="st">"upbmAux"</span>)</span>
<span id="cb11-2"><a href="#cb11-2"></a><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(pbm_8x60k_v1<span class="op">@</span>design)</span></code></pre></div>
<pre><code>##   Column Row              probeID
## 1      1   1      GE_BrightCorner
## 2      2   1      GE_BrightCorner
## 3      3   1           DarkCorner
## 4      4   1           DarkCorner
## 5      5   1 Cbf_5b_Hi_n359_o2_r1
## 6      6   1      dBr_14334_Jan07
##                                                       Sequence
## 1                                                         #N/A
## 2                                                         #N/A
## 3                                                         #N/A
## 4                                                         #N/A
## 5 GAAGCTATTCAGATCGACGTGACATGTATATAGTAGGTCTGTGTTCCGTTGTCCGTGCTG
## 6 GGTGTGAGTCCATTTCGTCAAACCAACGCAACAGGTGTCTGTGTTCCGTTGTCCGTGCTG</code></pre>
<p>We split the data.frame by scan type to create separate Alexa488 and Cy3 <em>PBMExperiment</em> objects. Note that the following lines of code are not evaluated in this vignette since the corresponding GPR files are not included in this package.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a>hoxc9alexa &lt;-<span class="st"> </span><span class="kw"><a href="../reference/gpr2PBMExperiment.html">gpr2PBMExperiment</a></span>(<span class="dt">scans =</span> dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(hoxc9table, type <span class="op">==</span><span class="st"> "Alexa"</span>),</span>
<span id="cb13-2"><a href="#cb13-2"></a>                                <span class="dt">probes =</span> pbm_8x60k_v1)</span>
<span id="cb13-3"><a href="#cb13-3"></a>hoxc9cy3 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/gpr2PBMExperiment.html">gpr2PBMExperiment</a></span>(<span class="dt">scans =</span> dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(hoxc9table, type <span class="op">==</span><span class="st"> "Cy3"</span>),</span>
<span id="cb13-4"><a href="#cb13-4"></a>                                <span class="dt">probes =</span> pbm_8x60k_v1)</span></code></pre></div>
<p>After the Alexa488 and Cy3 GPR files have been read in as <em>PBMExperiment</em> objects, we perform basic pre-processing on the probe intensities of each assay. Preprocessing includes [1] background subtraction, [2] Cy3 normalization, [3] spatial adjustment, [4] within and [5] across replicate (plate) normalization. To illustrate these steps, we use the subsetted <em>PBMExperiment</em> objects, <code>alexa_subset</code> and <code>cy3_subset</code>. While each pre-processing step can be run separately, a single wrapper function, <code><a href="../reference/upbmPreprocess.html">upbmPreprocess(..)</a></code> is provided to quickly perform all pre-processing steps on properly formatted PBM data. Details on each step is provided in a separate vignette. Here, we run all steps with the default parameters.</p>
<p>To perform Cy3 normalization, in addition to the Cy3 scan data, we need a reference distribution of expected Cy3 intensities (generated by <code><a href="../reference/cy3GenerateRef.html">cy3GenerateRef(..)</a></code>). A reference set generated using 90 Cy3 scans is included in the <em>upbmAux</em> package.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a><span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(refcy3_8x60k_v1, <span class="dt">package =</span> <span class="st">"upbmAux"</span>)</span></code></pre></div>
<p>Using this reference distribution, we perform pre-processing. By default, Alexa488 and Cy3 scans specified to <code><a href="../reference/upbmPreprocess.html">upbmPreprocess(..)</a></code> are expected to contain colData columns named <code>id_idx</code> which can be used to map scans across the two <em>PBMExperiment</em> objects. While the <code>id_idx</code> column may include repeated values in the Alexa488 data, the values must be unique for the Cy3 data. Additionally, the colData of the Alexa488 <em>PBMExperiment</em> is expected to have two additional columns, <code>id</code> and <code>condition</code>, corresponding to replicate (plate) and condition IDs, respectively. These labels are needed to perform within and between replicate normalization. Additionally, a reference condition is expected to exist in the <code>condition</code> colData column. The reference condition is identified by checking for conditions ending in <code>-REF</code>. These parameters may all be modified to specify different reference conditions and colData columns.</p>
<p>By default <code><a href="../reference/upbmPreprocess.html">upbmPreprocess(..)</a></code> will provide a weakly verbose log of the normalization steps.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a>alexa_normed &lt;-<span class="st"> </span><span class="kw"><a href="../reference/upbmPreprocess.html">upbmPreprocess</a></span>(alexa_subset, cy3_subset, <span class="dt">cy3refpe =</span> refcy3_8x60k_v1)</span></code></pre></div>
<pre><code>## || upbm::upbmPreprocess 
## || - Starting PBM data preprocessing.
## || - 6 Alexa488 PBM scans.
## || - 6 Cy3 PBM scans.
## ||
## || Background intensity subtraction ... 
## || ... finished!
## ||
## || Cy3 normalization using [ cy3FitEmpirical ] ... 
## || ... finished!
## ||
## || Spatial adjustment ... 
## || ... finished!
## ||
## || Within-replicate normalization ... 
## || ... finished!
## ||
## || Cross-replicate normalization ... 
## || ... finished!
## ||
## || Finished PBM data preprocessing.
## || Returning PBMExperiment with 41944 rows and 6 columns.</code></pre>
<p>The returned object is a <em>PBMExperiment</em> with the same name number of columns as the original Alexa488 <em>PBMExperiment</em> object. The number of rows is decreased to only include probes designated as signal probes based on filtering criteria specified as part of the corresponding <em>PBMDesign</em>.</p>
<p>After normalization, probe-level intensities are aggregated across replicates by fitting a probe-level model. Again, a <code>condition</code> column is expected to exist in the colData of the <em>PBMExperiment</em> object.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a>alexa_pfit &lt;-<span class="st"> </span><span class="kw"><a href="../reference/probeFit.html">probeFit</a></span>(alexa_normed)</span></code></pre></div>
<p>A <em>PBExperiment</em> object is returned with columns corresponding to the unique values of the colData <code>condition</code> column (here, HOXC9-R222W, HOXC9-REF).</p>
<p>Next, the probe-level summaries aggregated across replicates are used to obtain K-mer-level summaries. A helper function, <code><a href="../reference/uniqueKmers.html">uniqueKmers(..)</a></code> is provided for generating the full set of K-mers of a specified length (unique to reverse complementation, i.e. only one of <code>AAAACCCC</code> and <code>GGGGTTTT</code> is included). Performing K-mer inference on the full set of 8-mers can take close to a minute for each condition. While this is not a large amount of time, it is too long for the purpose of illustrating the function in this vignette. Therefore, <strong>while not recommended in practice</strong>, we fit the K-mer-level summaries for only a subset of randomly selected K-mers.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(1L)</span>
<span id="cb18-2"><a href="#cb18-2"></a>kmer_subset &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="kw"><a href="../reference/uniqueKmers.html">uniqueKmers</a></span>(8L), <span class="fl">1e4</span>)</span>
<span id="cb18-3"><a href="#cb18-3"></a></span>
<span id="cb18-4"><a href="#cb18-4"></a>alexa_kfit &lt;-<span class="st"> </span><span class="kw"><a href="../reference/kmerFit.html">kmerFit</a></span>(alexa_pfit, <span class="dt">kmers =</span> kmer_subset) <span class="co">## for illustration only, see above</span></span></code></pre></div>
<p>The result is a <em>SummarizedExperiment</em> with K-mer-level affinity and variance estimates for each condition, with rows now corresponding to K-mers rather than probes.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a>alexa_kfit</span></code></pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 10000 2 
## metadata(1): baseline
## assays(5): affinityEstimate affinityVariance contrastDifference
##   contrastAverage contrastVariance
## rownames: NULL
## rowData names(1): seq
## colnames(2): HOXC9-R222W HOXC9-REF
## colData names(0):</code></pre>
<p>Finally, these k-mer affinities can be used to perform K-mer-level inference. To test for differential 8-mer affinity across conditions, between a baseline condition (e.g. the wild type HOXC9 TF) and all other conditions, we make a call to <code><a href="../reference/kmerTestContrast.html">kmerTestContrast(..)</a></code>. This step should be relatively quick since estimates of affinity and variance have already been computed in the previous step. Note, if a different baseline condition should be used, this should be specified to the earlier <code>kmerFit</code> function. Other methods for inference are also defined in <code><a href="../reference/kmerTestAffinity.html">kmerTestAffinity(..)</a></code> and <code><a href="../reference/kmerTestSpecificity.html">kmerTestSpecificity(..)</a></code>.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a>alexa_kdiff &lt;-<span class="st"> </span><span class="kw"><a href="../reference/kmerTestContrast.html">kmerTestContrast</a></span>(alexa_kfit)</span></code></pre></div>
<p>The resulting object is again a <em>SummarizedExperiment</em> with assays containing the results of differential affinity testing.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a>alexa_kdiff</span></code></pre></div>
<pre><code>## class: SummarizedExperiment 
## dim: 10000 2 
## metadata(0):
## assays(6): contrastAverage contrastDifference ... contrastP
##   contrastQ
## rownames: NULL
## rowData names(1): seq
## colnames(2): HOXC9-R222W HOXC9-REF
## colData names(0):</code></pre>
<p>K-mer level averages and differences across conditions are included in the <code>contrastAverage</code> and <code>contrastDifference</code> assays. FDR-controlling adjusted p-values are included in the <code>contrastQ</code> assay. We can tidy the results using <code><a href="https://rdrr.io/pkg/broom/man/reexports.html">broom::tidy</a></code> as described above, and visualize the results as an MA plot.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a><span class="kw"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">"ggplot2"</span>))</span>
<span id="cb24-2"><a href="#cb24-2"></a></span>
<span id="cb24-3"><a href="#cb24-3"></a>diffdat &lt;-<span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">left_join</a></span>(broom<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/broom/man/reexports.html">tidy</a></span>(alexa_kdiff, <span class="st">"contrastAverage"</span>, <span class="dt">long =</span> <span class="ot">TRUE</span>),</span>
<span id="cb24-4"><a href="#cb24-4"></a>                            broom<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/broom/man/reexports.html">tidy</a></span>(alexa_kdiff, <span class="st">"contrastDifference"</span>, <span class="dt">long =</span> <span class="ot">TRUE</span>),</span>
<span id="cb24-5"><a href="#cb24-5"></a>                            <span class="dt">by =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"seq"</span>, <span class="st">"cname"</span>), <span class="dt">suffix =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">".A"</span>, <span class="st">".M"</span>))</span>
<span id="cb24-6"><a href="#cb24-6"></a>diffdat &lt;-<span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">left_join</a></span>(diffdat,</span>
<span id="cb24-7"><a href="#cb24-7"></a>                            broom<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/broom/man/reexports.html">tidy</a></span>(alexa_kdiff, <span class="st">"contrastQ"</span>, <span class="dt">long =</span> <span class="ot">TRUE</span>),</span>
<span id="cb24-8"><a href="#cb24-8"></a>                            <span class="dt">by =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"seq"</span>, <span class="st">"cname"</span>))</span>
<span id="cb24-9"><a href="#cb24-9"></a>diffdat &lt;-<span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(diffdat, <span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(value))</span>
<span id="cb24-10"><a href="#cb24-10"></a>diffdat &lt;-<span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">rename</a></span>(diffdat, <span class="dt">adj.pval =</span> value)</span>
<span id="cb24-11"><a href="#cb24-11"></a></span>
<span id="cb24-12"><a href="#cb24-12"></a><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(diffdat, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> value.A, <span class="dt">y =</span> value.M, <span class="dt">color =</span> adj.pval <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.001</span>)) <span class="op">+</span></span>
<span id="cb24-13"><a href="#cb24-13"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>(<span class="dt">alpha =</span> <span class="dv">1</span><span class="op">/</span><span class="dv">10</span>) <span class="op">+</span></span>
<span id="cb24-14"><a href="#cb24-14"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span>(<span class="dt">yintercept =</span> <span class="dv">0</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb24-15"><a href="#cb24-15"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_color_brewer</a></span>(<span class="dt">palette =</span> <span class="st">"Set1"</span>, <span class="dt">direction =</span> <span class="dv">-1</span>) <span class="op">+</span></span>
<span id="cb24-16"><a href="#cb24-16"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() <span class="op">+</span></span>
<span id="cb24-17"><a href="#cb24-17"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/guides.html">guides</a></span>(<span class="dt">color =</span> <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html">guide_legend</a></span>(<span class="dt">override.aes =</span> <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">alpha =</span> <span class="dv">1</span>))) <span class="op">+</span></span>
<span id="cb24-18"><a href="#cb24-18"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span>(. <span class="op">~</span><span class="st"> </span>cname) <span class="op">+</span></span>
<span id="cb24-19"><a href="#cb24-19"></a><span class="st">    </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span>(<span class="st">"HOXC9 differential 8-mer affinity MA plot"</span>)</span></code></pre></div>
<p><img src="upbm_files/figure-html/quick-plot-res-1.png" width="576"></p>
</div>
<div id="related-work" class="section level1">
<h1 class="hasAnchor">
<a href="#related-work" class="anchor"></a>Related Work</h1>
<p>A separate set of tools developed in Perl for the analysis of universal-design PBMs (uPBMs) is available for download at the <a href="the_brain.bwh.harvard.edu/PBMAnalysisSuite/indexSep2017.html"><em>Universal PBM Analysis Suite</em> site</a> <span class="citation">(Berger and Bulyk 2009)</span>. The <em>upbm</em> package builds on the work of the <em>Universal PBM Analysis Suite</em>, and provides alternative summaries of uPBM data not developed as part of the original suite of tools.</p>
<p>A primary purpose of both software packages is the summarization of DNA binding affinities at the level of individual K-mers. To this end, the <em>Universal PBM Analysis Suite</em> summarizes K-mer binding affinity for individual PBM samples with two metrics: a non-parametric rank-based enrichment score (E-score), and the cross-probe median intensity (calculated across all probes containing the corresponding K-mer). In contrast, with <em>upbm</em>, K-mer binding affinities are estimated using a mixed effects model, with affinities effectively estimated as the weighted average log-transformed probe intensity across all probes containing the corresponding K-mer. This differs from the cross-probe median intensity in serveral important ways. First, the relative contribution of individual probes to the K-mer summary is inversely related to the variance of the probe across samples, downweighting highly variable probes. Second, to estimate probe-level variability, replicate samples are necessary. Third, because the variance structure is explicitly modeled, statistical testing and inference is possible at the level of K-mers.</p>
<p>In addition to the summarization of DNA binding affinities at the level of individual K-mers, the <em>upbm</em> package also allows for testing for differential affinity between groups, e.g. to contrast binding affinity across allelic variants of the same transcription factor. Additionally, the package also allows for identifying preferentially bound K-mers for each groups of replicate samples. The <em>Universal PBM Analysis Suite</em> provides the ability to construct mononucleotide position weight matrices (PWMs), again, for individual PBM samples. Approaches for testing differential affinity between samples using the output of the <em>Universal PBM Analysis Suite</em> have also been previously described <span class="citation">(Jiang, Liu, and Bulyk 2013; Barrera et al. 2016)</span>.</p>
<p>Many of the PBM pre-processing steps of the <em>Universal PBM Analysis Suite</em> are maintained in the <em>upbm</em> package. These steps have been re-implemented, with some modifications, e.g. in the process of Cy3 normalization, to enable and streamline the complete analysis of PBM data in <em>R</em>.</p>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references hanging-indent">
<div id="ref-barrera2016">
<p>Barrera, Luis A, Anastasia Vedenko, Jesse V Kurland, Julia M Rogers, Stephen S Gisselbrecht, Elizabeth J Rossin, Jaie Woodard, et al. 2016. “Survey of Variation in Human Transcription Factors Reveals Prevalent Dna Binding Changes.” <em>Science</em> 351 (6280): 1450–4.</p>
</div>
<div id="ref-berger2009">
<p>Berger, Michael F, and Martha L Bulyk. 2009. “Universal Protein-Binding Microarrays for the Comprehensive Characterization of the Dna-Binding Specificities of Transcription Factors.” <em>Nature Protocols</em> 4 (3): 393–411.</p>
</div>
<div id="ref-berger2006">
<p>Berger, Michael F, Anthony A Philippakis, Aaron M Qureshi, Fangxue S He, Preston W Estep III, and Martha L Bulyk. 2006. “Compact, Universal Dna Microarrays to Comprehensively Determine Transcription-Factor Binding Site Specificities.” <em>Nature Biotechnology</em> 24 (11): 1429–35.</p>
</div>
<div id="ref-jiang2013">
<p>Jiang, Bo, Jun S Liu, and Martha L Bulyk. 2013. “Bayesian Hierarchical Model of Protein-Binding Microarray K-Mer Data Reduces Noise and Identifies Transcription Factor Subclasses and Preferred K-Mers.” <em>Bioinformatics</em> 29 (11): 1390–8.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li>
<a href="#hoxc9-dataset">HOXC9 Dataset</a>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#illustrative-subset">Illustrative Subset</a></li>
      </ul>
</li>
      <li><a href="#tidy-data">Tidy Data</a></li>
      <li><a href="#quickstart">Quickstart</a></li>
      <li><a href="#related-work">Related Work</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Patrick Kimes.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
