<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Normalize across replicates — normalizeAcrossReplicates • upbm</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Normalize across replicates — normalizeAcrossReplicates" />
<meta property="og:description" content="Universal PBM experiments are often performed with several conditions of interest,
e.g. allelic variants, assayed on separate arrays of the same plate with few replicates.
Within and across plates, probe intensities can vary for biologically
uninteresting reasons, such as concentration differences. To explicitly correct for these
differences, normalization is performed in two steps.
First, normalization is performed within replicates (plates). More detail on this procedure
can be found in the normalizeWithinReplicates documentation.
Second, normalization is performed across replicates (plates) with the assumption that
biologically uninteresting differences between replicates affect probe intensities
both multiplicatively and additively on the log-scale. A single log-scale multiplicative normalization
factor is first estimated for all samples within a replicate. Then, a log-scale additive normalization
is estimated such that the median intensities of the baseline samples in each replicate
are equal. More details on this calculation are provided below." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">upbm</a>
        <span class="version label label-info" data-toggle="tooltip" data-placement="bottom" title="GitHub pre-release">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/upbm-quickstart.html">Quick Start</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">General</li>
    <li>
      <a href="../articles/upbm-quickstart.html">Quickstart Guide</a>
    </li>
    <li>
      <a href="../articles/upbm-classes.html">PBM Class Details</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Analysis</li>
    <li>
      <a href="../articles/analysis-preprocessing.html">Sample Pre-Processing</a>
    </li>
    <li>
      <a href="../articles/analysis-summarization.html">Probe and K-mer Summarization</a>
    </li>
    <li>
      <a href="../articles/analysis-inference.html">K-mer-Level Inference</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Other</li>
    <li>
      <a href="../articles/other-qcplots.html">Quality Control Plots</a>
    </li>
    <li>
      <a href="../articles/other-tidydata.html">Tidy PBM Data</a>
    </li>
  </ul>
</li>
<li>
  <a href="https://github.com/pkimes/upbm/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Normalize across replicates</h1>
    <small class="dont-index">Source: <a href='https://github.com/pkimes/upbm/blob/master/R/normalizeAcrossReplicates.R'><code>R/normalizeAcrossReplicates.R</code></a></small>
    <div class="hidden name"><code>normalizeAcrossReplicates.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Universal PBM experiments are often performed with several conditions of interest,
e.g. allelic variants, assayed on separate arrays of the same plate with few replicates.
Within and across plates, probe intensities can vary for biologically
uninteresting reasons, such as concentration differences. To explicitly correct for these
differences, normalization is performed in two steps.</p>
<p>First, normalization is performed within replicates (plates). More detail on this procedure
can be found in the <code><a href='normalizeWithinReplicates.html'>normalizeWithinReplicates</a></code> documentation.</p>
<p>Second, normalization is performed across replicates (plates) with the assumption that
biologically uninteresting differences between replicates affect probe intensities
both multiplicatively and additively on the log-scale. A single log-scale multiplicative normalization
factor is first estimated for all samples within a replicate. Then, a log-scale additive normalization
is estimated such that the median intensities of the <code>baseline</code> samples in each replicate
are equal. More details on this calculation are provided below.</p>
    </div>

    <pre class="usage"><span class='fu'>normalizeAcrossReplicates</span>(
  <span class='no'>pe</span>,
  <span class='kw'>assay</span> <span class='kw'>=</span> <span class='kw pkg'>SummarizedExperiment</span><span class='kw ns'>::</span><span class='fu'><a href='https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html'>assayNames</a></span>(<span class='no'>pe</span>)[<span class='fl'>1</span>],
  <span class='kw'>group</span> <span class='kw'>=</span> <span class='st'>"id"</span>,
  <span class='kw'>stratify</span> <span class='kw'>=</span> <span class='st'>"condition"</span>,
  <span class='kw'>baseline</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>pe</th>
      <td><p>SummarizedExperiment object containing GPR intensity information.</p></td>
    </tr>
    <tr>
      <th>assay</th>
      <td><p>a string name of the assay to normalize.
(default = <code>SummarizedExperiment::assayNames(pe)[1]</code>)</p></td>
    </tr>
    <tr>
      <th>group</th>
      <td><p>a character string specifying a column in <code>colData(pe)</code> to use for grouping replicates.
(default = <code>"id"</code>)</p></td>
    </tr>
    <tr>
      <th>stratify</th>
      <td><p>a character string specifying a column in <code>colData(pe)</code> to use for determining
the unique baseline scan within each <code>group</code> and to match samples across values of <code>group</code>.
(default = <code>"condition"</code>)</p></td>
    </tr>
    <tr>
      <th>baseline</th>
      <td><p>a character string specifying the baseline condition in the <code>stratify</code> column to normalize
other conditions against within each <code>group</code>. If not specified and set to NULL, the baseline
value is guessed by looking for values in the <code>stratify</code> column ending in ``ref". If multiple
unique matching values are found, a warning is thrown and the first matching sample is used.
(default = NULL)</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>a logical value whether to print verbose output during analysis. (default = FALSE)</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Original PBMExperiment object with assay containing cross-replicate normalized intensities
(<code>"normalized"</code>) and new columns added to the colData, <code>"acrossRepMultScale"</code> and <code>"acrossRepAddScale"</code>,
containing the inverse of the log-scale multiplicative and additive scaling factors used to normalize intensities.
If an assay with the same name is already included in the object, it will be overwritten.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The following procedure is used to estimate the log-scale multiplicative factor
for each replicate. First, a cross-replicate reference is computed for each baseline condition
(specified by <code>stratify=</code> and <code>baseline=</code>) by taking the cross-replicate mean quantiles 
of the observed log2 intensities. Next, a <em>per-replicate</em> log multiplicative scaling factor is
computed by taking the median ratio of the rank-ordered and median-centered log-probe intensities
between the baseline samples in each replicate and the reference distribution. Visually,
this can be interpreted as the approximate slope of the quantile-quantile (QQ) plot generated
using log-scale intensities. To reduce the impact of outlier probes, scaling factors are estimated
using only the middle 80<!-- % of probe intensities. --></p>
<p>After log-scale multiplicative factors have been estimated to correct for differences in
log-scale variance across replicates, a second log-scale additive factor is estimated
for each replicate to correct for differences in log-scale shift. A "global median" intensity
is first calculated across replicates by taking the geometric mean of the median
intensities in all <code>baseline</code> samples across replicates. This "global median" is computed
using the input probe intensities, i.e. without any cross-replicate normalization. 
The log-scale additive factor estimated as the difference between the median normalized probe
intensity of the <code>baseline</code> sample in each replicate and the "global median".
While the log-scale additive factor is estimated using only <code>baseline</code> samples, the normalization
is applied to all samples in the replicate.</p>
<p>Cross-replicate normalization is first carried out for replicates containing a baseline sample as
described above. Replicates without a baseline sample are then normalized to already normalized
replicates using overlapping conditions in the <code>stratify=</code> column.</p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://www.pkimes.com'>Patrick Kimes</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


