<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Fit probe models — probeFit • upbm</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Fit probe models — probeFit" />
<meta property="og:description" content="Given a PBMExperiment of normalized probe intensities, this function fits
robust probe-level models across samples using the lmFit and eBayes
functions from the limma package. The output is a PBMExperiment object with
probes in rows and conditions, rather than individual samples, in columns. The
condition labels of each sample should be specified as a column in the colData of
input PBMExperiment using the stratify= parameter. By default, the function
expects a &quot;condition&quot; column in the colData of the input object." />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">upbm</a>
        <span class="version label label-info" data-toggle="tooltip" data-placement="bottom" title="GitHub pre-release">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/upbm-quickstart.html">Quick Start</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">General</li>
    <li>
      <a href="../articles/upbm-quickstart.html">Quickstart Guide</a>
    </li>
    <li>
      <a href="../articles/upbm-classes.html">PBM Class Details</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Analysis</li>
    <li>
      <a href="../articles/analysis-preprocessing.html">Sample Pre-Processing</a>
    </li>
    <li>
      <a href="../articles/analysis-summarization.html">Probe and K-mer Summarization</a>
    </li>
    <li>
      <a href="../articles/analysis-inference.html">K-mer-Level Inference</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Other</li>
    <li>
      <a href="../articles/other-qcplots.html">Quality Control Plots</a>
    </li>
    <li>
      <a href="../articles/other-tidydata.html">Tidy PBM Data</a>
    </li>
  </ul>
</li>
<li>
  <a href="https://github.com/pkimes/upbm">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Fit probe models</h1>
    <small class="dont-index">Source: <a href='https://github.com/pkimes/upbm/blob/master/R/probeFit.R'><code>R/probeFit.R</code></a></small>
    <div class="hidden name"><code>probeFit.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Given a PBMExperiment of normalized probe intensities, this function fits
robust probe-level models across samples using the <code>lmFit</code> and <code>eBayes</code>
functions from the <span class="pkg">limma</span> package. The output is a PBMExperiment object with
probes in rows and conditions, rather than individual samples, in columns. The
condition labels of each sample should be specified as a column in the colData of
input PBMExperiment using the <code>stratify=</code> parameter. By default, the function
expects a <code>"condition"</code> column in the colData of the input object.</p>
    </div>

    <pre class="usage"><span class='fu'>probeFit</span>(<span class='no'>pe</span>, <span class='kw'>assay</span> <span class='kw'>=</span> <span class='kw pkg'>SummarizedExperiment</span><span class='kw ns'>::</span><span class='fu'><a href='https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html'>assayNames</a></span>(<span class='no'>pe</span>)[<span class='fl'>1</span>],
  <span class='kw'>stratify</span> <span class='kw'>=</span> <span class='st'>"condition"</span>, <span class='kw'>guardrail</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>offset</span> <span class='kw'>=</span> <span class='fl'>1L</span>,
  <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='no'>...</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>pe</th>
      <td><p>a PBMExperiment object containing PBM intensity data.</p></td>
    </tr>
    <tr>
      <th>assay</th>
      <td><p>a string name of the assay to use for fitting probe models.
(default = <code>SummarizedExperiment::assayNames(pe)[1]</code>)</p></td>
    </tr>
    <tr>
      <th>stratify</th>
      <td><p>a character string specifying a column in <code>colData(pe)</code> to
use for grouping samples, e.g. into alleles. (default = <code>"condition"</code>)</p></td>
    </tr>
    <tr>
      <th>guardrail</th>
      <td><p>a logical value whether to stop function if any sample appears
to be of clearly low quality. Currently, this only checks for whether
more than 20% of probes in any sample are NA. If any criteria is met,
an error will be returned. (default = TRUE)</p></td>
    </tr>
    <tr>
      <th>offset</th>
      <td><p>an integer offset to add to intensities before log2 scaling to
prevent errors with zero intensities. If set to 0, probes with
zero intensities are dropped/ignored in estimation. (default = 1)</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>a logical value whether to print verbose output during analysis. (default = FALSE)</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>additional parameters to be passed to <code><a href='https://rdrr.io/pkg/limma/man/ebayes.html'>limma::eBayes</a></code>. See
Details for default parameters used in this function as they differ from
the <code><a href='https://rdrr.io/pkg/limma/man/ebayes.html'>limma::eBayes</a></code> default values.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>PBMExperiment object with probe-level log2 intensity information aggregated
across replicates for each unique value in <code>stratify</code>. Columns in the returned
PBMExperiment correspond to unique values of the <code>stratify</code> variable, and
rows correspond to the subset of probes passing the filtering criteria
of the original PBMExperiment object. The PBMExperiment includes three assays,
<code>"beta"</code>, <code>"sd"</code>, and <code>"df"</code>, storing the cross-replicate
mean, standard deviation, and degrees of freedom estimates for each condition and probe.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>More specifically, <code><a href='https://rdrr.io/pkg/limma/man/lmFit.html'>limma::lmFit</a></code> is called on the log2-transformed values of the
specified assay with no intercept and a binary design matrix with columns corresponding
to the unique values of <code>colData(pe)[[stratify]]</code>. Since PBM experiments are often
performed with few replicates per condition, the fit is then passed to <code><a href='https://rdrr.io/pkg/limma/man/ebayes.html'>limma::eBayes</a></code>
with <code>robust = TRUE</code> and <code>trend = TRUE</code> by default. Parameters to
<code><a href='https://rdrr.io/pkg/limma/man/ebayes.html'>limma::eBayes</a></code> can be modified by specifying them during the call to <code>probeFit</code>.</p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>The cross-probe robust and trended variance estimates are obtained using the
<code>eBayes</code> function from the limma package with <code>robust = TRUE</code> and
<code>trend = TRUE</code> by default. The procedure was developed and described
in the following paper:</p><ul>
<li><p>Phipson, B., Lee, S., Majewski, I. J., Alexander, W. S., &amp; Smyth, G. K. (2016). Robust hyperparameter estimation protects against hypervariable genes and improves power to detect differential expression. Annals of Applied Statistics, 10(2), 946-963.</p></li>
</ul>


  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
      <li><a href="#details">Details</a></li>
      <li><a href="#references">References</a></li>
    </ul>

    <h2>Author</h2>
    <p>Patrick Kimes</p>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://www.pkimes.com'>Patrick Kimes</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


